<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Reading, Editing and Writing DNA Sequences with Ape • blaseRtools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reading, Editing and Writing DNA Sequences with Ape">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9197</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a></li>
    <li><a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a></li>
    <li><a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SummarizedHeatmap.html">Summarized Heatmap Vignette</a></li>
    <li><a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reading, Editing and Writing DNA Sequences with Ape</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/vignettes/Ape.Rmd" class="external-link"><code>vignettes/Ape.Rmd</code></a></small>
      <div class="d-none name"><code>Ape.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attach the packages you will need for the analysis.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/blaserlab/blaseRtools" class="external-link">blaseRtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/blaserlab/blaseRdata" class="external-link">blaseRdata</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></code></pre></div>
<p>This is a brief tutorial to show you how to programmatically read and
write DNA sequences in a user-friendly format.</p>
<p>This set of functions are designed to work with genbank-formatted DNA
sequence files. Genbank is a long-standing, text-based DNA sequence
format. There are many programs to view and edit genbank-format files.
My recommendation is that you use <a href="https://jorgensen.biology.utah.edu/wayned/ape/" class="external-link">ApE</a>. This
program is very easy to use.</p>
<div class="section level2">
<h2 id="ape-class">Ape Class<a class="anchor" aria-label="anchor" href="#ape-class"></a>
</h2>
<p>We designed a data structure which we refer to as “Ape”, in reference
to the gui-based program mentioned above, which is designed to hold
genbank-formatted DNA sequence data within an R programming
environment.</p>
<p>The motivation behind this is as follows: Let’s say you wanted to
clone a gene enhancer from genomic DNA, or edit a specific exon with
CRISPR, or map putative transcription factor binding sites. One way to
do this is to go to ensembl, download the sequence, manually annotate
the sequence files as desired, and then work on your experiment. This is
100% interactive and 0% programmatic. You can also use pure
R/bioconductor tools to get sequence and feature information. This is
more programmatic, but the problem is that you will have to learn how to
access genomic sequence data via R and use GenomicRanges and Biostrings
to define your features. Some of this could be done programmatically but
it will require a lot of hard coding to get it done. This is not how I
would want to accomplish this task.</p>
<p>The Ape class and few associated functions automate most of these
tasks for you. You can read and write genebank/Ape format files for
interactive and programmatic editing. Ape class objects store sequence
internally as a <a href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html" class="external-link">Biostring</a>
and features as a <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html" class="external-link">GenomicRanges</a>
object so you can use those interfaces for your data as desired.</p>
<p>In the companion blaseRdata package we have precompiled hg38 and
danRer11 genome references and gene models for human and zebrafish so
this is one-stop shopping for sequence data. Mouse data will be added in
a future version.</p>
</div>
<div class="section level2">
<h2 id="reading-creating-or-writing-genbankape-files">Reading, Creating, or Writing Genbank/Ape Files<a class="anchor" aria-label="anchor" href="#reading-creating-or-writing-genbankape-files"></a>
</h2>
<p>Lets say you have a collection of genbank/Ape files that you would
like to read into R so that you can extract the sequence and/or feature
information for programmatic analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in the data</span></span>
<span></span>
<span><span class="va">vignette_CXCL8_ape</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_parseape.html">bb_parseape</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/hg38_CXCL8.ape"</span>, package <span class="op">=</span> <span class="st">"blaseRdata"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>You can show the sequence information in your R console by typing the
name of the Ape object like so:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Show the Ape Data </span></span>
<span></span>
<span><span class="va">vignette_CXCL8_ape</span></span>
<span><span class="co">#&gt; LOCUS       CXCL8                   3176 bp ds-DNA     linear        21-JAN-2022  </span></span>
<span><span class="co">#&gt; DEFINITION     </span></span>
<span><span class="co">#&gt; ACCESSION     </span></span>
<span><span class="co">#&gt; VERSION     </span></span>
<span><span class="co">#&gt; SOURCE     </span></span>
<span><span class="co">#&gt; COMMENT     Sequence is from Homo sapiens, GRCh38.</span></span>
<span><span class="co">#&gt; COMMENT     Gene models are from ensembl.</span></span>
<span><span class="co">#&gt; COMMENT     Gene is CXCL8</span></span>
<span><span class="co">#&gt; COMMENT     Extensions 0 bp left and 0 bp right.</span></span>
<span><span class="co">#&gt; COMMENT     Final genomic coordinates are:</span></span>
<span><span class="co">#&gt; COMMENT     chr4 73740541-73743716</span></span>
<span><span class="co">#&gt; FEATURES             Location/Qualifiers</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      gene            1..3176</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_gene"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             119..182</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             1002..1137</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             1409..1496</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             119..182</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             1002..1137</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             1409..1492</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             1909..1924</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      mRNA            1..1790</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_mRNA"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      five_prime_UTR  1..118</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_five_prime_UTR"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#9ecae1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#a1d99b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1..182</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_exon_1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1002..1137</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_exon_4"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1409..1790</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_exon_7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      three_prime_UTR 1497..1790</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_three_prime_UTR"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#9ecae1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#a1d99b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      lnc_RNA         29..1446</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_lnc_RNA"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            29..182</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_exon_2"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1002..1446</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_exon_5"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      mRNA            29..3176</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_mRNA"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      five_prime_UTR  29..118</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_five_prime_UTR"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#9ecae1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#a1d99b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            29..182</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_exon_3"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1002..1137</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_exon_6"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1409..1492</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_exon_8"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1909..3176</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_exon_9"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      three_prime_UTR 1925..3176</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL8_three_prime_UTR"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#9ecae1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#a1d99b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ORIGIN</span></span>
<span><span class="co">#&gt;         1 AAAAAGCCAC CGGAGCACTC CATAAGGCAC AAACTTTCAG AGACAGCAGA GCACACAAGC</span></span>
<span><span class="co">#&gt;        61 TTCTAGGACA AGAGCCAGGA AGAAACCACC GGAAGGAACC ATCTCACTGT GTGTAAACAT</span></span>
<span><span class="co">#&gt;       121 GACTTCCAAG CTGGCCGTGG CTCTCTTGGC AGCCTTCCTG ATTTCTGCAG CTCTGTGTGA</span></span>
<span><span class="co">#&gt;       181 AGGTAAGCAC ATCTTTCTGA CCTACAGCGT TTTCCTATGT CTAAATGTGA TCCTTAGATA</span></span>
<span><span class="co">#&gt;       241 GCAAAGCTAT TCTTGATGCT TTGGTAACAA ACATCCTTTT TATTCAGAAA CAGAATATAA</span></span>
<span><span class="co">#&gt;       301 TCTTAGCAGT CAATTAATGT TAAATTGAAG ATTTAGAAAA AACTATATAT AACACTTAGG</span></span>
<span><span class="co">#&gt;       361 AAAGTATAAA GTTTGATCAA TATAGATATT CTGCTTTTAT AATTTATACC ATGTAGCATG</span></span>
<span><span class="co">#&gt;       421 CATATATTTA ACGTAAATAA GTAATTTATA GTATGTCCTA TTGAGAACCA CGGTTACCTA</span></span>
<span><span class="co">#&gt;       481 TATTATGTAT TAATATTGAG TTGAGCAAGG TAACTCAGAC AATTCCACTC CTTGTAGTAT</span></span>
<span><span class="co">#&gt;       541 TTCATTGACA AGCCTCAGAT TTGTCATTAA TTCCTGTCTG GTTTAAAGAT ACCCTGATTA</span></span>
<span><span class="co">#&gt;       601 TAGACCAGGC ATGTATAACT TATTTATATA TTTCTGTTAA TTCTTTCTGA AGGCAATTTC</span></span>
<span><span class="co">#&gt;       661 TATGCTGGAG AGTCTTAGCT TGCCTACTAT AAATAACACT GTGGTATCAC AGAGGATTAT</span></span>
<span><span class="co">#&gt;       721 GCAATATTGA CCAGATAAAA ATACCATGAA GATGTTGATA TTGTACAAAA AGAACTCTAA</span></span>
<span><span class="co">#&gt;       781 CTCTTTATAT AGGAAGTCGT TCAATGTTGT CAGTTATGAC TGTTTTTTAA AACAAAGAAC</span></span>
<span><span class="co">#&gt;       841 TAACTGAGGT CAAGGGCTAG GAGAATATTC AGGAATGAGT TCACTAGAAA CATGATGCCT</span></span>
<span><span class="co">#&gt;       901 TCCATAGTCT CCAAATAATC ATATTGGAAT TAGAAAGGAA GTAGCTGGCA GAGCTGTGCC</span></span>
<span><span class="co">#&gt;       961 TGTTGATAAA ATCAATCCTT AATCACTTTT TCCCCCAACA GGTGCAGTTT TGCCAAGGAG</span></span>
<span><span class="co">#&gt;      1021 TGCTAAAGAA CTTAGATGTC AGTGCATAAA GACATACTCC AAACCTTTCC ACCCCAAATT</span></span>
<span><span class="co">#&gt;      1081 TATCAAAGAA CTGAGAGTGA TTGAGAGTGG ACCACACTGC GCCAACACAG AAATTATGTA</span></span>
<span><span class="co">#&gt;      1141 AGTACTTTAA AAAAGATTAG ATATTTTGTT TTAGCAAACT TAAAATTAAG GAAGGTGGAA</span></span>
<span><span class="co">#&gt;      1201 ATATTTAGGA AAGTTCCAGG TGTTAGGATT ACAGTAGTAA ATGAAACAAA ACAAAATAAA</span></span>
<span><span class="co">#&gt;      1261 AATATTTGTC TACATGACAT TTAAATATGG TAGCTTCCAC AACTACTATA AATGTTATTT</span></span>
<span><span class="co">#&gt;      1321 TGGACTTAGA CTTTATGCCT GACTTAAGGA ATCATGATTT GAATGCAAAA ACTAAATATT</span></span>
<span><span class="co">#&gt;      1381 AATCTGAACC ATTTCTTTCT TATTTCAGTG TAAAGCTTTC TGATGGAAGA GAGCTCTGTC</span></span>
<span><span class="co">#&gt;      1441 TGGACCCCAA GGAAAACTGG GTGCAGAGGG TTGTGGAGAA GTTTTTGAAG AGGTAAGTTA</span></span>
<span><span class="co">#&gt;      1501 TATATTTTTT AATTTAAATT TTTCATTTAT CCTGAGACAT ATAATCCAAA GTCAGCCTAT</span></span>
<span><span class="co">#&gt;      1561 AAATTTCTTT CTGTTGCTAA AAATCGTCAT TAGGTATCTG CCTTTTTGGT TAAAAAAAAA</span></span>
<span><span class="co">#&gt;      1621 AGGAATAGCA TCAATAGTGA GTTTGTTGTA CTCATGACCA GAAAGACCAT ACATAGTTTG</span></span>
<span><span class="co">#&gt;      1681 CCCAGGAAAT TCTGGGTTTA AGCTTGTGTC CTATACTCTT AGTAAAGTTC TTTGTCACTC</span></span>
<span><span class="co">#&gt;      1741 CCAGTAGTGT CCTATTTTAG ATGATAATTT CTTTGATCTC CCTATTTATA GTTGAGAATA</span></span>
<span><span class="co">#&gt;      1801 TAGAGCATTT CTAACACATG AATGTCAAAG ACTATATTGA CTTTTCAAGA ACCCTACTTT</span></span>
<span><span class="co">#&gt;      1861 CCTTCTTATT AAACATAGCT CATCTTTATA TTTTTAATTT TATTTTAGGG CTGAGAATTC</span></span>
<span><span class="co">#&gt;      1921 ATAAAAAAAT TCATTCTCTG TGGTATCCAA GAATCAGTGA AGATGCCAGT GAAACTTCAA</span></span>
<span><span class="co">#&gt;      1981 GCAAATCTAC TTCAACACTT CATGTATTGT GTGGGTCTGT TGTAGGGTTG CCAGATGCAA</span></span>
<span><span class="co">#&gt;      2041 TACAAGATTC CTGGTTAAAT TTGAATTTCA GTAAACAATG AATAGTTTTT CATTGTACCA</span></span>
<span><span class="co">#&gt;      2101 TGAAATATCC AGAACATACT TATATGTAAA GTATTATTTA TTTGAATCTA CAAAAAACAA</span></span>
<span><span class="co">#&gt;      2161 CAAATAATTT TTAAATATAA GGATTTTCCT AGATATTGCA CGGGAGAATA TACAAATAGC</span></span>
<span><span class="co">#&gt;      2221 AAAATTGAGG CCAAGGGCCA AGAGAATATC CGAACTTTAA TTTCAGGAAT TGAATGGGTT</span></span>
<span><span class="co">#&gt;      2281 TGCTAGAATG TGATATTTGA AGCATCACAT AAAAATGATG GGACAATAAA TTTTGCCATA</span></span>
<span><span class="co">#&gt;      2341 AAGTCAAATT TAGCTGGAAA TCCTGGATTT TTTTCTGTTA AATCTGGCAA CCCTAGTCTG</span></span>
<span><span class="co">#&gt;      2401 CTAGCCAGGA TCCACAAGTC CTTGTTCCAC TGTGCCTTGG TTTCTCCTTT ATTTCTAAGT</span></span>
<span><span class="co">#&gt;      2461 GGAAAAAGTA TTAGCCACCA TCTTACCTCA CAGTGATGTT GTGAGGACAT GTGGAAGCAC</span></span>
<span><span class="co">#&gt;      2521 TTTAAGTTTT TTCATCATAA CATAAATTAT TTTCAAGTGT AACTTATTAA CCTATTTATT</span></span>
<span><span class="co">#&gt;      2581 ATTTATGTAT TTATTTAAGC ATCAAATATT TGTGCAAGAA TTTGGAAAAA TAGAAGATGA</span></span>
<span><span class="co">#&gt;      2641 ATCATTGATT GAATAGTTAT AAAGATGTTA TAGTAAATTT ATTTTATTTT AGATATTAAA</span></span>
<span><span class="co">#&gt;      2701 TGATGTTTTA TTAGATAAAT TTCAATCAGG GTTTTTAGAT TAAACAAACA AACAATTGGG</span></span>
<span><span class="co">#&gt;      2761 TACCCAGTTA AATTTTCATT TCAGATAAAC AACAAATAAT TTTTTAGTAT AAGTACATTA</span></span>
<span><span class="co">#&gt;      2821 TTGTTTATCT GAAATTTTAA TTGAACTAAC AATCCTAGTT TGATACTCCC AGTCTTGTCA</span></span>
<span><span class="co">#&gt;      2881 TTGCCAGCTG TGTTGGTAGT GCTGTGTTGA ATTACGGAAT AATGAGTTAG AACTATTAAA</span></span>
<span><span class="co">#&gt;      2941 ACAGCCAAAA CTCCACAGTC AATATTAGTA ATTTCTTGCT GGTTGAAACT TGTTTATTAT</span></span>
<span><span class="co">#&gt;      3001 GTACAAATAG ATTCTTATAA TATTATTTAA ATGACTGCAT TTTTAAATAC AAGGCTTTAT</span></span>
<span><span class="co">#&gt;      3061 ATTTTTAACT TTAAGATGTT TTTATGTGCT CTCCAAATTT TTTTTACTGT TTCTGATTGT</span></span>
<span><span class="co">#&gt;      3121 ATGGAAATAT AAAAGTAAAT ATGAAACATT TAAAATATAA TTTGTTGTCA AAGTAA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; //</span></span></code></pre></div>
<p>Printing this data out to the console may not be the most useful
thing to do, but it shows you that everything has been parsed correctly.
It should look like a genbank/Ape file.</p>
<p>You can also create an Ape object based on hg38 or danRer11 reference
data like so:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">vignette_CXCL1_ape</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_make_ape_genomic.html">bb_make_ape_genomic</a></span><span class="op">(</span><span class="st">"CXCL1"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vignette_CXCL1_ape</span></span>
<span><span class="co">#&gt; LOCUS       CXCL1                   1916 bp ds-DNA     linear        19-DEC-2025 </span></span>
<span><span class="co">#&gt; COMMENT     Sequence is from Homo sapiens hg38.</span></span>
<span><span class="co">#&gt; COMMENT     Gene models are from ensembl.</span></span>
<span><span class="co">#&gt; COMMENT     Gene is CXCL1</span></span>
<span><span class="co">#&gt; COMMENT     Extensions 0 bp left and 0 bp right.</span></span>
<span><span class="co">#&gt; COMMENT     Final genomic coordinates are:</span></span>
<span><span class="co">#&gt; COMMENT     chr4 73869393-73871308</span></span>
<span><span class="co">#&gt; FEATURES             Location/Qualifiers</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      gene            1..1916</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_gene"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             79..178</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             277..400</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             514..597</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             1129..1144</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      lnc_RNA         1..785</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_lnc_RNA"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1..178</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            277..400</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_3"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            514..785</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_5"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      mRNA            1..1916</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_mRNA"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      five_prime_UTR  1..78</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_five_prime_UTR"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#9ecae1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#a1d99b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1..178</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_2"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            277..400</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_4"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            514..597</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_6"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1129..1916</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      three_prime_UTR 1145..1916</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_three_prime_UTR"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#9ecae1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#a1d99b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ORIGIN</span></span>
<span><span class="co">#&gt;         1 ACAGAGCCCG GGCCGCAGGC ACCTCCTCGC CAGCTCTTCC GCTCCTCTCA CAGCCGCCAG</span></span>
<span><span class="co">#&gt;        61 ACCCGCCTGC TGAGCCCCAT GGCCCGCGCT GCTCTCTCCG CCGCCCCCAG CAATCCCCGG</span></span>
<span><span class="co">#&gt;       121 CTCCTGCGAG TGGCACTGCT GCTCCTGCTC CTGGTAGCCG CTGGCCGGCG CGCAGCAGGT</span></span>
<span><span class="co">#&gt;       181 GGGTACCGGC GCCCTGGGGT CCCCGGGCCG GACGCGGCTG GGGTAGGCAC CCAGCGCCGA</span></span>
<span><span class="co">#&gt;       241 CAGCCTCGCT CAGTCAGTGA GTCTCTTCTT CCCTAGGAGC GTCCGTGGCC ACTGAACTGC</span></span>
<span><span class="co">#&gt;       301 GCTGCCAGTG CTTGCAGACC CTGCAGGGAA TTCACCCCAA GAACATCCAA AGTGTGAACG</span></span>
<span><span class="co">#&gt;       361 TGAAGTCCCC CGGACCCCAC TGCGCCCAAA CCGAAGTCAT GTAAGTCCCG CCCCGCGCTG</span></span>
<span><span class="co">#&gt;       421 CCTCTGCCAC CGCCGGGGTC CCAGACCCTC CTGCTGCCCC AACCCTGTCC CCAGCCCGAC</span></span>
<span><span class="co">#&gt;       481 CTCCTGCCTC ACGAGATTCC CTTCCCTCTG CAGAGCCACA CTCAAGAATG GGCGGAAAGC</span></span>
<span><span class="co">#&gt;       541 TTGCCTCAAT CCTGCATCCC CCATAGTTAA GAAAATCATC GAAAAGATGC TGAACAGGTG</span></span>
<span><span class="co">#&gt;       601 AGTTATGGTT TCCATGTACA CAGGCGACTG GAGCCGTTGG TCAGAAATAC TGGCATGTGC</span></span>
<span><span class="co">#&gt;       661 CCCCTAAAAA TAAAATCAGG AAAACCCAGG GGTTAGTTGA AGGACTAGAA ATTGGGATTA</span></span>
<span><span class="co">#&gt;       721 TTGTTTTCAC AATTAAGGTT TCCTTTACGA TAATTACTGC TCTGGTGCCA GAGGATATTC</span></span>
<span><span class="co">#&gt;       781 CCAATGCCTG GCGTCCCCAC CCTGGTTCTT CCTTCGTTCC AATGAATGTA GGTAAAACTG</span></span>
<span><span class="co">#&gt;       841 CCTTCATTTG AGGCCCAGTA GGACAAACAG CAACAGGTTC TGGCTGTTTT TAATCCAATA</span></span>
<span><span class="co">#&gt;       901 GTACAGTGGA GACCACCGCC CCACCCCACC CCCATTCCTA AAAGAGCATC CCAAGCTTAG</span></span>
<span><span class="co">#&gt;       961 AGGTCCCTGC CACACAGCAC AGCTGTCATA GGCAGTAGCC ACTTGGTTGC CAGGCTGGGG</span></span>
<span><span class="co">#&gt;      1021 AAACTGCATT CGGAAAACTC TAGAGGCTGG AGGAGCAGGG CAGGAGAAGA GTGTTGTGCA</span></span>
<span><span class="co">#&gt;      1081 ATCAGCTTTC CCGAGCACCT ACTCAGGGCA CCCATTTTCT CATTGCAGTG ACAAATCCAA</span></span>
<span><span class="co">#&gt;      1141 CTGACCAGAA GGGAGGAGGA AGCTCACTGG TGGCTGTTCC TGAAGGAGGC CCTGCCCTTA</span></span>
<span><span class="co">#&gt;      1201 TAGGAACAGA AGAGGAAAGA GAGACACAGC TGCAGAGGCC ACCTGGATTG TGCCTAATGT</span></span>
<span><span class="co">#&gt;      1261 GTTTGAGCAT CGCTTAGGAG AAGTCTTCTA TTTATTTATT TATTCATTAG TTTTGAAGAT</span></span>
<span><span class="co">#&gt;      1321 TCTATGTTAA TATTTTAGGT GTAAAATAAT TAAGGGTATG ATTAACTCTA CCTGCACACT</span></span>
<span><span class="co">#&gt;      1381 GTCCTATTAT ATTCATTCTT TTTGAAATGT CAACCCCAAG TTAGTTCAAT CTGGATTCAT</span></span>
<span><span class="co">#&gt;      1441 ATTTAATTTG AAGGTAGAAT GTTTTCAAAT GTTCTCCAGT CATTATGTTA ATATTTCTGA</span></span>
<span><span class="co">#&gt;      1501 GGAGCCTGCA ACATGCCAGC CACTGTGATA GAGGCTGGCG GATCCAAGCA AATGGCCAAT</span></span>
<span><span class="co">#&gt;      1561 GAGATCATTG TGAAGGCAGG GGAATGTATG TGCACATCTG TTTTGTAACT GTTTAGATGA</span></span>
<span><span class="co">#&gt;      1621 ATGTCAGTTG TTATTTATTG AAATGATTTC ACAGTGTGTG GTCAACATTT CTCATGTTGA</span></span>
<span><span class="co">#&gt;      1681 AACTTTAAGA ACTAAAATGT TCTAAATATC CCTTGGACAT TTTATGTCTT TCTTGTAAGG</span></span>
<span><span class="co">#&gt;      1741 CATACTGCCT TGTTTAATGG TAGTTTTACA GTGTTTCTGG CTTAGAACAA AGGGGCTTAA</span></span>
<span><span class="co">#&gt;      1801 TTATTGATGT TTTCATAGAG AATATAAAAA TAAAGCACTT ATAGAAAAAA CTCGTTTGAT</span></span>
<span><span class="co">#&gt;      1861 TTTTGGGGGG AAACAAGGGC TACCTTTACT GGAAAATCTG GTGATTTATA TCAATA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; //</span></span></code></pre></div>
<p>You can select certain feature types to include by specifying the
<code>include_type</code> argument.</p>
<p>You may also add flanking sequences or specify genomic regions by
coordinate rather than gene name.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get genomic sequence and extend 100 bp left and right.</span></span>
<span><span class="va">vignette_CXCL1_ape</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_make_ape_genomic.html">bb_make_ape_genomic</a></span><span class="op">(</span><span class="st">"CXCL1"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, extend_left <span class="op">=</span> <span class="fl">100</span>, extend_right <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vignette_CXCL1_ape</span></span>
<span><span class="co">#&gt; LOCUS       CXCL1                   2116 bp ds-DNA     linear        19-DEC-2025 </span></span>
<span><span class="co">#&gt; COMMENT     Sequence is from Homo sapiens hg38.</span></span>
<span><span class="co">#&gt; COMMENT     Gene models are from ensembl.</span></span>
<span><span class="co">#&gt; COMMENT     Gene is CXCL1</span></span>
<span><span class="co">#&gt; COMMENT     Extensions 100 bp left and 100 bp right.</span></span>
<span><span class="co">#&gt; COMMENT     Final genomic coordinates are:</span></span>
<span><span class="co">#&gt; COMMENT     chr4 73869293-73871408</span></span>
<span><span class="co">#&gt; FEATURES             Location/Qualifiers</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      gene            101..2016</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_gene"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             179..278</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             377..500</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             614..697</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             1229..1244</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      lnc_RNA         101..885</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_lnc_RNA"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            101..278</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            377..500</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_3"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            614..885</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_5"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      mRNA            101..2016</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_mRNA"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      five_prime_UTR  101..178</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_five_prime_UTR"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#9ecae1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#a1d99b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            101..278</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_2"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            377..500</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_4"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            614..697</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_6"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1229..2016</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      three_prime_UTR 1245..2016</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_three_prime_UTR"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#9ecae1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#a1d99b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ORIGIN</span></span>
<span><span class="co">#&gt;         1 TCGGGATCGA TCTGGAACTC CGGGAATTTC CCTGGCCCGG GGGCTCCGGG CTTTCCAGCC</span></span>
<span><span class="co">#&gt;        61 CCAACCATGC ATAAAAGGGG TTCGCGGATC TCGGAGAGCC ACAGAGCCCG GGCCGCAGGC</span></span>
<span><span class="co">#&gt;       121 ACCTCCTCGC CAGCTCTTCC GCTCCTCTCA CAGCCGCCAG ACCCGCCTGC TGAGCCCCAT</span></span>
<span><span class="co">#&gt;       181 GGCCCGCGCT GCTCTCTCCG CCGCCCCCAG CAATCCCCGG CTCCTGCGAG TGGCACTGCT</span></span>
<span><span class="co">#&gt;       241 GCTCCTGCTC CTGGTAGCCG CTGGCCGGCG CGCAGCAGGT GGGTACCGGC GCCCTGGGGT</span></span>
<span><span class="co">#&gt;       301 CCCCGGGCCG GACGCGGCTG GGGTAGGCAC CCAGCGCCGA CAGCCTCGCT CAGTCAGTGA</span></span>
<span><span class="co">#&gt;       361 GTCTCTTCTT CCCTAGGAGC GTCCGTGGCC ACTGAACTGC GCTGCCAGTG CTTGCAGACC</span></span>
<span><span class="co">#&gt;       421 CTGCAGGGAA TTCACCCCAA GAACATCCAA AGTGTGAACG TGAAGTCCCC CGGACCCCAC</span></span>
<span><span class="co">#&gt;       481 TGCGCCCAAA CCGAAGTCAT GTAAGTCCCG CCCCGCGCTG CCTCTGCCAC CGCCGGGGTC</span></span>
<span><span class="co">#&gt;       541 CCAGACCCTC CTGCTGCCCC AACCCTGTCC CCAGCCCGAC CTCCTGCCTC ACGAGATTCC</span></span>
<span><span class="co">#&gt;       601 CTTCCCTCTG CAGAGCCACA CTCAAGAATG GGCGGAAAGC TTGCCTCAAT CCTGCATCCC</span></span>
<span><span class="co">#&gt;       661 CCATAGTTAA GAAAATCATC GAAAAGATGC TGAACAGGTG AGTTATGGTT TCCATGTACA</span></span>
<span><span class="co">#&gt;       721 CAGGCGACTG GAGCCGTTGG TCAGAAATAC TGGCATGTGC CCCCTAAAAA TAAAATCAGG</span></span>
<span><span class="co">#&gt;       781 AAAACCCAGG GGTTAGTTGA AGGACTAGAA ATTGGGATTA TTGTTTTCAC AATTAAGGTT</span></span>
<span><span class="co">#&gt;       841 TCCTTTACGA TAATTACTGC TCTGGTGCCA GAGGATATTC CCAATGCCTG GCGTCCCCAC</span></span>
<span><span class="co">#&gt;       901 CCTGGTTCTT CCTTCGTTCC AATGAATGTA GGTAAAACTG CCTTCATTTG AGGCCCAGTA</span></span>
<span><span class="co">#&gt;       961 GGACAAACAG CAACAGGTTC TGGCTGTTTT TAATCCAATA GTACAGTGGA GACCACCGCC</span></span>
<span><span class="co">#&gt;      1021 CCACCCCACC CCCATTCCTA AAAGAGCATC CCAAGCTTAG AGGTCCCTGC CACACAGCAC</span></span>
<span><span class="co">#&gt;      1081 AGCTGTCATA GGCAGTAGCC ACTTGGTTGC CAGGCTGGGG AAACTGCATT CGGAAAACTC</span></span>
<span><span class="co">#&gt;      1141 TAGAGGCTGG AGGAGCAGGG CAGGAGAAGA GTGTTGTGCA ATCAGCTTTC CCGAGCACCT</span></span>
<span><span class="co">#&gt;      1201 ACTCAGGGCA CCCATTTTCT CATTGCAGTG ACAAATCCAA CTGACCAGAA GGGAGGAGGA</span></span>
<span><span class="co">#&gt;      1261 AGCTCACTGG TGGCTGTTCC TGAAGGAGGC CCTGCCCTTA TAGGAACAGA AGAGGAAAGA</span></span>
<span><span class="co">#&gt;      1321 GAGACACAGC TGCAGAGGCC ACCTGGATTG TGCCTAATGT GTTTGAGCAT CGCTTAGGAG</span></span>
<span><span class="co">#&gt;      1381 AAGTCTTCTA TTTATTTATT TATTCATTAG TTTTGAAGAT TCTATGTTAA TATTTTAGGT</span></span>
<span><span class="co">#&gt;      1441 GTAAAATAAT TAAGGGTATG ATTAACTCTA CCTGCACACT GTCCTATTAT ATTCATTCTT</span></span>
<span><span class="co">#&gt;      1501 TTTGAAATGT CAACCCCAAG TTAGTTCAAT CTGGATTCAT ATTTAATTTG AAGGTAGAAT</span></span>
<span><span class="co">#&gt;      1561 GTTTTCAAAT GTTCTCCAGT CATTATGTTA ATATTTCTGA GGAGCCTGCA ACATGCCAGC</span></span>
<span><span class="co">#&gt;      1621 CACTGTGATA GAGGCTGGCG GATCCAAGCA AATGGCCAAT GAGATCATTG TGAAGGCAGG</span></span>
<span><span class="co">#&gt;      1681 GGAATGTATG TGCACATCTG TTTTGTAACT GTTTAGATGA ATGTCAGTTG TTATTTATTG</span></span>
<span><span class="co">#&gt;      1741 AAATGATTTC ACAGTGTGTG GTCAACATTT CTCATGTTGA AACTTTAAGA ACTAAAATGT</span></span>
<span><span class="co">#&gt;      1801 TCTAAATATC CCTTGGACAT TTTATGTCTT TCTTGTAAGG CATACTGCCT TGTTTAATGG</span></span>
<span><span class="co">#&gt;      1861 TAGTTTTACA GTGTTTCTGG CTTAGAACAA AGGGGCTTAA TTATTGATGT TTTCATAGAG</span></span>
<span><span class="co">#&gt;      1921 AATATAAAAA TAAAGCACTT ATAGAAAAAA CTCGTTTGAT TTTTGGGGGG AAACAAGGGC</span></span>
<span><span class="co">#&gt;      1981 TACCTTTACT GGAAAATCTG GTGATTTATA TCAATATTTC TCAATTTTTT AATTGTGTTT</span></span>
<span><span class="co">#&gt;      2041 ATTTTTCTGG GTGTTCAATT TGCTATACAG ATAAATCAAA CTATGAGTTA TGCTTCATTT</span></span>
<span><span class="co">#&gt;      2101 CATGCGGTGA TTGCTG</span></span>
<span><span class="co">#&gt; //</span></span></code></pre></div>
<p>Note that the extension arguments are strictly relative to the top or
+ strand, so the sense of upstream or downstream may be different
relative to your gene of interest if that gene is on the top or bottom
strand.</p>
<p>You will notice that all feature coordinates have been recalculated
and are relative to the sequence present in the object. Let’s say you
had a feature that you knew about in terms of genomic coordinates that
you wanted include when making the object. Here is how you do that:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get genomic sequence and extend 100 bp left and right.</span></span>
<span><span class="co"># Now add a new custom feature based on original coordinates:  chr4 73869293-73871408</span></span>
<span><span class="va">vignette_CXCL1_ape</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_make_ape_genomic.html">bb_make_ape_genomic</a></span><span class="op">(</span></span>
<span>  <span class="st">"CXCL1"</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  extend_left <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  extend_right <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  additional_granges <span class="op">=</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html" class="external-link">makeGRangesFromDataFrame</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>      seqname <span class="op">=</span> <span class="st">"chr4"</span>,</span>
<span>      start <span class="op">=</span> <span class="fl">73869293</span>,</span>
<span>      end <span class="op">=</span> <span class="fl">73871408</span>,</span>
<span>      strand <span class="op">=</span> <span class="st">"+"</span>,</span>
<span>      gene_name <span class="op">=</span> <span class="st">"CXCL1"</span>,</span>
<span>      type <span class="op">=</span> <span class="st">"custom_feature"</span>,</span>
<span>      label <span class="op">=</span> <span class="st">"custom_feature_1"</span>,</span>
<span>      fwdcolor <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>      revcolor <span class="op">=</span> <span class="st">"blue"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    keep.extra.columns <span class="op">=</span> <span class="cn">T</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">vignette_CXCL1_ape</span></span>
<span><span class="co">#&gt; LOCUS       CXCL1                   2116 bp ds-DNA     linear        19-DEC-2025 </span></span>
<span><span class="co">#&gt; COMMENT     Sequence is from Homo sapiens hg38.</span></span>
<span><span class="co">#&gt; COMMENT     Gene models are from ensembl.</span></span>
<span><span class="co">#&gt; COMMENT     Gene is CXCL1</span></span>
<span><span class="co">#&gt; COMMENT     Extensions 100 bp left and 100 bp right.</span></span>
<span><span class="co">#&gt; COMMENT     Final genomic coordinates are:</span></span>
<span><span class="co">#&gt; COMMENT     chr4 73869293-73871408</span></span>
<span><span class="co">#&gt; FEATURES             Location/Qualifiers</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      gene            101..2016</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_gene"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             179..278</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             377..500</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             614..697</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      CDS             1229..1244</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_CDS"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      lnc_RNA         101..885</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_lnc_RNA"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            101..278</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            377..500</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_3"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            614..885</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_5"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      mRNA            101..2016</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_mRNA"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      five_prime_UTR  101..178</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_five_prime_UTR"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#9ecae1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#a1d99b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            101..278</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_2"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            377..500</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_4"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            614..697</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_6"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      exon            1229..2016</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_exon_7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#3182bd"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#31a354</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      three_prime_UTR 1245..2016</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_three_prime_UTR"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#9ecae1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#a1d99b</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      custom_feature  1..2116</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_custom_feature_1"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="red"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="blue</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ORIGIN</span></span>
<span><span class="co">#&gt;         1 TCGGGATCGA TCTGGAACTC CGGGAATTTC CCTGGCCCGG GGGCTCCGGG CTTTCCAGCC</span></span>
<span><span class="co">#&gt;        61 CCAACCATGC ATAAAAGGGG TTCGCGGATC TCGGAGAGCC ACAGAGCCCG GGCCGCAGGC</span></span>
<span><span class="co">#&gt;       121 ACCTCCTCGC CAGCTCTTCC GCTCCTCTCA CAGCCGCCAG ACCCGCCTGC TGAGCCCCAT</span></span>
<span><span class="co">#&gt;       181 GGCCCGCGCT GCTCTCTCCG CCGCCCCCAG CAATCCCCGG CTCCTGCGAG TGGCACTGCT</span></span>
<span><span class="co">#&gt;       241 GCTCCTGCTC CTGGTAGCCG CTGGCCGGCG CGCAGCAGGT GGGTACCGGC GCCCTGGGGT</span></span>
<span><span class="co">#&gt;       301 CCCCGGGCCG GACGCGGCTG GGGTAGGCAC CCAGCGCCGA CAGCCTCGCT CAGTCAGTGA</span></span>
<span><span class="co">#&gt;       361 GTCTCTTCTT CCCTAGGAGC GTCCGTGGCC ACTGAACTGC GCTGCCAGTG CTTGCAGACC</span></span>
<span><span class="co">#&gt;       421 CTGCAGGGAA TTCACCCCAA GAACATCCAA AGTGTGAACG TGAAGTCCCC CGGACCCCAC</span></span>
<span><span class="co">#&gt;       481 TGCGCCCAAA CCGAAGTCAT GTAAGTCCCG CCCCGCGCTG CCTCTGCCAC CGCCGGGGTC</span></span>
<span><span class="co">#&gt;       541 CCAGACCCTC CTGCTGCCCC AACCCTGTCC CCAGCCCGAC CTCCTGCCTC ACGAGATTCC</span></span>
<span><span class="co">#&gt;       601 CTTCCCTCTG CAGAGCCACA CTCAAGAATG GGCGGAAAGC TTGCCTCAAT CCTGCATCCC</span></span>
<span><span class="co">#&gt;       661 CCATAGTTAA GAAAATCATC GAAAAGATGC TGAACAGGTG AGTTATGGTT TCCATGTACA</span></span>
<span><span class="co">#&gt;       721 CAGGCGACTG GAGCCGTTGG TCAGAAATAC TGGCATGTGC CCCCTAAAAA TAAAATCAGG</span></span>
<span><span class="co">#&gt;       781 AAAACCCAGG GGTTAGTTGA AGGACTAGAA ATTGGGATTA TTGTTTTCAC AATTAAGGTT</span></span>
<span><span class="co">#&gt;       841 TCCTTTACGA TAATTACTGC TCTGGTGCCA GAGGATATTC CCAATGCCTG GCGTCCCCAC</span></span>
<span><span class="co">#&gt;       901 CCTGGTTCTT CCTTCGTTCC AATGAATGTA GGTAAAACTG CCTTCATTTG AGGCCCAGTA</span></span>
<span><span class="co">#&gt;       961 GGACAAACAG CAACAGGTTC TGGCTGTTTT TAATCCAATA GTACAGTGGA GACCACCGCC</span></span>
<span><span class="co">#&gt;      1021 CCACCCCACC CCCATTCCTA AAAGAGCATC CCAAGCTTAG AGGTCCCTGC CACACAGCAC</span></span>
<span><span class="co">#&gt;      1081 AGCTGTCATA GGCAGTAGCC ACTTGGTTGC CAGGCTGGGG AAACTGCATT CGGAAAACTC</span></span>
<span><span class="co">#&gt;      1141 TAGAGGCTGG AGGAGCAGGG CAGGAGAAGA GTGTTGTGCA ATCAGCTTTC CCGAGCACCT</span></span>
<span><span class="co">#&gt;      1201 ACTCAGGGCA CCCATTTTCT CATTGCAGTG ACAAATCCAA CTGACCAGAA GGGAGGAGGA</span></span>
<span><span class="co">#&gt;      1261 AGCTCACTGG TGGCTGTTCC TGAAGGAGGC CCTGCCCTTA TAGGAACAGA AGAGGAAAGA</span></span>
<span><span class="co">#&gt;      1321 GAGACACAGC TGCAGAGGCC ACCTGGATTG TGCCTAATGT GTTTGAGCAT CGCTTAGGAG</span></span>
<span><span class="co">#&gt;      1381 AAGTCTTCTA TTTATTTATT TATTCATTAG TTTTGAAGAT TCTATGTTAA TATTTTAGGT</span></span>
<span><span class="co">#&gt;      1441 GTAAAATAAT TAAGGGTATG ATTAACTCTA CCTGCACACT GTCCTATTAT ATTCATTCTT</span></span>
<span><span class="co">#&gt;      1501 TTTGAAATGT CAACCCCAAG TTAGTTCAAT CTGGATTCAT ATTTAATTTG AAGGTAGAAT</span></span>
<span><span class="co">#&gt;      1561 GTTTTCAAAT GTTCTCCAGT CATTATGTTA ATATTTCTGA GGAGCCTGCA ACATGCCAGC</span></span>
<span><span class="co">#&gt;      1621 CACTGTGATA GAGGCTGGCG GATCCAAGCA AATGGCCAAT GAGATCATTG TGAAGGCAGG</span></span>
<span><span class="co">#&gt;      1681 GGAATGTATG TGCACATCTG TTTTGTAACT GTTTAGATGA ATGTCAGTTG TTATTTATTG</span></span>
<span><span class="co">#&gt;      1741 AAATGATTTC ACAGTGTGTG GTCAACATTT CTCATGTTGA AACTTTAAGA ACTAAAATGT</span></span>
<span><span class="co">#&gt;      1801 TCTAAATATC CCTTGGACAT TTTATGTCTT TCTTGTAAGG CATACTGCCT TGTTTAATGG</span></span>
<span><span class="co">#&gt;      1861 TAGTTTTACA GTGTTTCTGG CTTAGAACAA AGGGGCTTAA TTATTGATGT TTTCATAGAG</span></span>
<span><span class="co">#&gt;      1921 AATATAAAAA TAAAGCACTT ATAGAAAAAA CTCGTTTGAT TTTTGGGGGG AAACAAGGGC</span></span>
<span><span class="co">#&gt;      1981 TACCTTTACT GGAAAATCTG GTGATTTATA TCAATATTTC TCAATTTTTT AATTGTGTTT</span></span>
<span><span class="co">#&gt;      2041 ATTTTTCTGG GTGTTCAATT TGCTATACAG ATAAATCAAA CTATGAGTTA TGCTTCATTT</span></span>
<span><span class="co">#&gt;      2101 CATGCGGTGA TTGCTG</span></span>
<span><span class="co">#&gt; //</span></span></code></pre></div>
<p>This turns out to be a useful feature but it does force you into
using some GenomicRanges code which is verbose and less familiar. The
help page for this function will remind you how to create the GRanges
object. Note that you can supply any number of GRanges in the object to
generate features.</p>
<p>Methods are provided for saving Ape objects as either genbank/Ape
files or fasta files:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Save as a genbank/Ape file</span></span>
<span><span class="fu"><a href="../reference/Ape.save.html">Ape.save</a></span><span class="op">(</span><span class="va">vignette_CXCL1_ape</span>, out <span class="op">=</span> <span class="st">"/path/to/file/filename.ape"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save as fasta</span></span>
<span><span class="fu"><a href="../reference/Ape.fasta.html">Ape.fasta</a></span><span class="op">(</span><span class="va">vignette_CXCL1_ape</span>, out <span class="op">=</span> <span class="st">"/path/to/file/filename.fa"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="getting-and-setting-ape-object-slots">Getting and Setting Ape Object Slots<a class="anchor" aria-label="anchor" href="#getting-and-setting-ape-object-slots"></a>
</h2>
<p>You may want to programmatically change the features of an Ape
object, or extract sequences or features to use elsewhere. Here you will
need to know a bit more about GRanges and Biostrings.</p>
<p>To get the sequence or features from an Ape Object:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the sequence</span></span>
<span><span class="fu"><a href="../reference/Ape.DNA.html">Ape.DNA</a></span><span class="op">(</span><span class="va">vignette_CXCL1_ape</span><span class="op">)</span></span>
<span><span class="co">#&gt; DNAStringSet object of length 1:</span></span>
<span><span class="co">#&gt;     width seq                                               names               </span></span>
<span><span class="co">#&gt; [1]  2116 <span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GGG</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFAF;">CC</span><span style="color: #000000; background-color: #AFFFFF;">GG</span>...<span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TTT</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #AFFFFF;">GG</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span> ape_seq</span></span>
<span></span>
<span><span class="co"># get the features</span></span>
<span><span class="fu"><a href="../reference/Ape.granges.html">Ape.granges</a></span><span class="op">(</span><span class="va">vignette_CXCL1_ape</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRanges object with 17 ranges and 4 metadata columns:</span></span>
<span><span class="co">#&gt;                          seqnames    ranges strand |            type</span></span>
<span><span class="co">#&gt;                             &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;</span></span>
<span><span class="co">#&gt;               CXCL1_gene  ape_seq  101-2016      + |            gene</span></span>
<span><span class="co">#&gt;                CXCL1_CDS  ape_seq   179-278      + |             CDS</span></span>
<span><span class="co">#&gt;                CXCL1_CDS  ape_seq   377-500      + |             CDS</span></span>
<span><span class="co">#&gt;                CXCL1_CDS  ape_seq   614-697      + |             CDS</span></span>
<span><span class="co">#&gt;                CXCL1_CDS  ape_seq 1229-1244      + |             CDS</span></span>
<span><span class="co">#&gt;                      ...      ...       ...    ... .             ...</span></span>
<span><span class="co">#&gt;             CXCL1_exon_4  ape_seq   377-500      + |            exon</span></span>
<span><span class="co">#&gt;             CXCL1_exon_6  ape_seq   614-697      + |            exon</span></span>
<span><span class="co">#&gt;             CXCL1_exon_7  ape_seq 1229-2016      + |            exon</span></span>
<span><span class="co">#&gt;    CXCL1_three_prime_UTR  ape_seq 1245-2016      + | three_prime_UTR</span></span>
<span><span class="co">#&gt;   CXCL1_custom_feature_1  ape_seq    1-2116      + |  custom_feature</span></span>
<span><span class="co">#&gt;                                       locus_tag    fwdcolor    revcolor</span></span>
<span><span class="co">#&gt;                                     &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;               CXCL1_gene             CXCL1_gene     #deebf7     #e5f5e0</span></span>
<span><span class="co">#&gt;                CXCL1_CDS              CXCL1_CDS     #deebf7     #e5f5e0</span></span>
<span><span class="co">#&gt;                CXCL1_CDS              CXCL1_CDS     #deebf7     #e5f5e0</span></span>
<span><span class="co">#&gt;                CXCL1_CDS              CXCL1_CDS     #deebf7     #e5f5e0</span></span>
<span><span class="co">#&gt;                CXCL1_CDS              CXCL1_CDS     #deebf7     #e5f5e0</span></span>
<span><span class="co">#&gt;                      ...                    ...         ...         ...</span></span>
<span><span class="co">#&gt;             CXCL1_exon_4           CXCL1_exon_4     #3182bd     #31a354</span></span>
<span><span class="co">#&gt;             CXCL1_exon_6           CXCL1_exon_6     #3182bd     #31a354</span></span>
<span><span class="co">#&gt;             CXCL1_exon_7           CXCL1_exon_7     #3182bd     #31a354</span></span>
<span><span class="co">#&gt;    CXCL1_three_prime_UTR  CXCL1_three_prime_UTR     #9ecae1     #a1d99b</span></span>
<span><span class="co">#&gt;   CXCL1_custom_feature_1 CXCL1_custom_feature_1         red        blue</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>You can set the features of an Ape object like so:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># define the new feature set</span></span>
<span><span class="va">old_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Ape.granges.html">Ape.granges</a></span><span class="op">(</span><span class="va">vignette_CXCL1_ape</span><span class="op">)</span></span>
<span><span class="va">new_features</span> <span class="op">&lt;-</span> <span class="va">old_features</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">old_features</span><span class="op">)</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"gene"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">new_vignette_CXCL1_ape</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Ape.setFeatures.html">Ape.setFeatures</a></span><span class="op">(</span><span class="va">vignette_CXCL1_ape</span>, gr <span class="op">=</span> <span class="va">new_features</span><span class="op">)</span></span>
<span><span class="va">new_vignette_CXCL1_ape</span></span>
<span><span class="co">#&gt; LOCUS       CXCL1                   2116 bp ds-DNA     linear        19-DEC-2025 </span></span>
<span><span class="co">#&gt; COMMENT     Sequence is from Homo sapiens hg38.</span></span>
<span><span class="co">#&gt; COMMENT     Gene models are from ensembl.</span></span>
<span><span class="co">#&gt; COMMENT     Gene is CXCL1</span></span>
<span><span class="co">#&gt; COMMENT     Extensions 100 bp left and 100 bp right.</span></span>
<span><span class="co">#&gt; COMMENT     Final genomic coordinates are:</span></span>
<span><span class="co">#&gt; COMMENT     chr4 73869293-73871408</span></span>
<span><span class="co">#&gt; FEATURES             Location/Qualifiers</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      gene            101..2016</span></span>
<span><span class="co">#&gt;                      /locus_tag="CXCL1_gene"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_fwdcolor="#deebf7"</span></span>
<span><span class="co">#&gt;                      /ApEinfo_revcolor="#e5f5e0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ORIGIN</span></span>
<span><span class="co">#&gt;         1 TCGGGATCGA TCTGGAACTC CGGGAATTTC CCTGGCCCGG GGGCTCCGGG CTTTCCAGCC</span></span>
<span><span class="co">#&gt;        61 CCAACCATGC ATAAAAGGGG TTCGCGGATC TCGGAGAGCC ACAGAGCCCG GGCCGCAGGC</span></span>
<span><span class="co">#&gt;       121 ACCTCCTCGC CAGCTCTTCC GCTCCTCTCA CAGCCGCCAG ACCCGCCTGC TGAGCCCCAT</span></span>
<span><span class="co">#&gt;       181 GGCCCGCGCT GCTCTCTCCG CCGCCCCCAG CAATCCCCGG CTCCTGCGAG TGGCACTGCT</span></span>
<span><span class="co">#&gt;       241 GCTCCTGCTC CTGGTAGCCG CTGGCCGGCG CGCAGCAGGT GGGTACCGGC GCCCTGGGGT</span></span>
<span><span class="co">#&gt;       301 CCCCGGGCCG GACGCGGCTG GGGTAGGCAC CCAGCGCCGA CAGCCTCGCT CAGTCAGTGA</span></span>
<span><span class="co">#&gt;       361 GTCTCTTCTT CCCTAGGAGC GTCCGTGGCC ACTGAACTGC GCTGCCAGTG CTTGCAGACC</span></span>
<span><span class="co">#&gt;       421 CTGCAGGGAA TTCACCCCAA GAACATCCAA AGTGTGAACG TGAAGTCCCC CGGACCCCAC</span></span>
<span><span class="co">#&gt;       481 TGCGCCCAAA CCGAAGTCAT GTAAGTCCCG CCCCGCGCTG CCTCTGCCAC CGCCGGGGTC</span></span>
<span><span class="co">#&gt;       541 CCAGACCCTC CTGCTGCCCC AACCCTGTCC CCAGCCCGAC CTCCTGCCTC ACGAGATTCC</span></span>
<span><span class="co">#&gt;       601 CTTCCCTCTG CAGAGCCACA CTCAAGAATG GGCGGAAAGC TTGCCTCAAT CCTGCATCCC</span></span>
<span><span class="co">#&gt;       661 CCATAGTTAA GAAAATCATC GAAAAGATGC TGAACAGGTG AGTTATGGTT TCCATGTACA</span></span>
<span><span class="co">#&gt;       721 CAGGCGACTG GAGCCGTTGG TCAGAAATAC TGGCATGTGC CCCCTAAAAA TAAAATCAGG</span></span>
<span><span class="co">#&gt;       781 AAAACCCAGG GGTTAGTTGA AGGACTAGAA ATTGGGATTA TTGTTTTCAC AATTAAGGTT</span></span>
<span><span class="co">#&gt;       841 TCCTTTACGA TAATTACTGC TCTGGTGCCA GAGGATATTC CCAATGCCTG GCGTCCCCAC</span></span>
<span><span class="co">#&gt;       901 CCTGGTTCTT CCTTCGTTCC AATGAATGTA GGTAAAACTG CCTTCATTTG AGGCCCAGTA</span></span>
<span><span class="co">#&gt;       961 GGACAAACAG CAACAGGTTC TGGCTGTTTT TAATCCAATA GTACAGTGGA GACCACCGCC</span></span>
<span><span class="co">#&gt;      1021 CCACCCCACC CCCATTCCTA AAAGAGCATC CCAAGCTTAG AGGTCCCTGC CACACAGCAC</span></span>
<span><span class="co">#&gt;      1081 AGCTGTCATA GGCAGTAGCC ACTTGGTTGC CAGGCTGGGG AAACTGCATT CGGAAAACTC</span></span>
<span><span class="co">#&gt;      1141 TAGAGGCTGG AGGAGCAGGG CAGGAGAAGA GTGTTGTGCA ATCAGCTTTC CCGAGCACCT</span></span>
<span><span class="co">#&gt;      1201 ACTCAGGGCA CCCATTTTCT CATTGCAGTG ACAAATCCAA CTGACCAGAA GGGAGGAGGA</span></span>
<span><span class="co">#&gt;      1261 AGCTCACTGG TGGCTGTTCC TGAAGGAGGC CCTGCCCTTA TAGGAACAGA AGAGGAAAGA</span></span>
<span><span class="co">#&gt;      1321 GAGACACAGC TGCAGAGGCC ACCTGGATTG TGCCTAATGT GTTTGAGCAT CGCTTAGGAG</span></span>
<span><span class="co">#&gt;      1381 AAGTCTTCTA TTTATTTATT TATTCATTAG TTTTGAAGAT TCTATGTTAA TATTTTAGGT</span></span>
<span><span class="co">#&gt;      1441 GTAAAATAAT TAAGGGTATG ATTAACTCTA CCTGCACACT GTCCTATTAT ATTCATTCTT</span></span>
<span><span class="co">#&gt;      1501 TTTGAAATGT CAACCCCAAG TTAGTTCAAT CTGGATTCAT ATTTAATTTG AAGGTAGAAT</span></span>
<span><span class="co">#&gt;      1561 GTTTTCAAAT GTTCTCCAGT CATTATGTTA ATATTTCTGA GGAGCCTGCA ACATGCCAGC</span></span>
<span><span class="co">#&gt;      1621 CACTGTGATA GAGGCTGGCG GATCCAAGCA AATGGCCAAT GAGATCATTG TGAAGGCAGG</span></span>
<span><span class="co">#&gt;      1681 GGAATGTATG TGCACATCTG TTTTGTAACT GTTTAGATGA ATGTCAGTTG TTATTTATTG</span></span>
<span><span class="co">#&gt;      1741 AAATGATTTC ACAGTGTGTG GTCAACATTT CTCATGTTGA AACTTTAAGA ACTAAAATGT</span></span>
<span><span class="co">#&gt;      1801 TCTAAATATC CCTTGGACAT TTTATGTCTT TCTTGTAAGG CATACTGCCT TGTTTAATGG</span></span>
<span><span class="co">#&gt;      1861 TAGTTTTACA GTGTTTCTGG CTTAGAACAA AGGGGCTTAA TTATTGATGT TTTCATAGAG</span></span>
<span><span class="co">#&gt;      1921 AATATAAAAA TAAAGCACTT ATAGAAAAAA CTCGTTTGAT TTTTGGGGGG AAACAAGGGC</span></span>
<span><span class="co">#&gt;      1981 TACCTTTACT GGAAAATCTG GTGATTTATA TCAATATTTC TCAATTTTTT AATTGTGTTT</span></span>
<span><span class="co">#&gt;      2041 ATTTTTCTGG GTGTTCAATT TGCTATACAG ATAAATCAAA CTATGAGTTA TGCTTCATTT</span></span>
<span><span class="co">#&gt;      2101 CATGCGGTGA TTGCTG</span></span>
<span><span class="co">#&gt; //</span></span></code></pre></div>
<p>That shows you how to subset a granges object. If you want to combine
two granges objects, run <code>c(object_1, object_2)</code> etc.</p>
<div class="section level3">
<h3 id="annotating-putative-transcription-factor-binding-sites">Annotating Putative Transcription Factor Binding Sites<a class="anchor" aria-label="anchor" href="#annotating-putative-transcription-factor-binding-sites"></a>
</h3>
<p>This is a pretty common function and there are several other tools
you can use to do this, but here we use <a href="https://meme-suite.org/meme/doc/fimo.html" class="external-link">FIMO</a> from the MEME
suite to do the annotation.</p>
<p>The prerequisite here is that you have to install FIMO on your
system. Instructions are available at the link provided. To find TFBS in
your Ape file you will use the <code><a href="../reference/Ape.fimo.html">Ape.fimo()</a></code> function.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Ape.fimo.html">Ape.fimo</a></span><span class="op">(</span><span class="va">vignette_CXCL1_ape</span>, fimo_feature <span class="op">=</span> <span class="st">"CXCL1_gene"</span><span class="op">)</span></span></code></pre></div>
<p>For this to run properly you have to identify the feature name within
the Ape object that you want to evaluate. This is the name of the Ape
object GRanges element or the locus_tag from the genbank/Ape file.</p>
<p>If your installation of FIMO can’t be found it will provide
instructions for you.</p>
<p>The function returns a GRanges object with the results. Optionally
(recommended) you can specify an output directory for detailed output
data.</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Good luck working with your DNA sequences!</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
