<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>scRNA-seq Data Analysis • blaseRtools</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="scRNA-seq Data Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9197</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a></li>
    <li><a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a></li>
    <li><a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SummarizedHeatmap.html">Summarized Heatmap Vignette</a></li>
    <li><a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>scRNA-seq Data Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/vignettes/scRNAseq.Rmd" class="external-link"><code>vignettes/scRNAseq.Rmd</code></a></small>
      <div class="d-none name"><code>scRNAseq.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette will show you how to load 10X single cell RNA seq data
into R and to perform “vanilla” processing. Here is an overview of the
steps we will cover:</p>
<ol style="list-style-type: decimal">
<li>Pre-processing data using 10X genomics cloud.</li>
<li>Loading data into a list of cellDataSet objects.</li>
<li>QC Functions</li>
<li>Merging into a single cellDataSet object.</li>
<li>Dimension reduction</li>
<li>Batch correction</li>
<li>Clustering</li>
<li>Gene modules</li>
<li>Label transfer using Seurat</li>
<li>Cell assignments</li>
<li>UMAP Plot Types</li>
<li>Gene Dotplots</li>
<li>Differential Gene Expression</li>
</ol>
<p>I use these steps on almost every scRNA-seq analysis I perform. With
proper configuration, steps 1-9 can be run non-interactively. I usually
will make a “dry run” interactively as I build the processing script.
Then I will refresh the R session and source the script to run from
beginning to end.</p>
<p>The last steps require some interactive input and may be customized
depending on the analysis you are doing. These usually become small
standalone scripts.</p>
<div class="section level3">
<h3 id="an-important-note-about-data-management">An important note about data management<a class="anchor" aria-label="anchor" href="#an-important-note-about-data-management"></a>
</h3>
<p>An important fact to note is that the way we run these algorithms
(UMAP dimension reduction and clustering in particular), the results are
subtly non-deterministic. By this I mean that there will be subtle,
arbitrary changes in the way the clusters are defined or the exact
dimensions assigned to particular cells. This does not change the
underlying expression data or affect the analysis negatively except that
if you were to run the same script twice, cluster 5 may become cluster 6
and UMAP dimensions may look different. The bottom line is that it is
important to know which steps are deterministic and which are not. Also,
which steps are computationally intensive and which are not.</p>
<p>In order to deal with these issues, there are particular data objects
which you will want to save as .rda files and then go back to if needed
in the future. Usually this will be the final cellDataSet object and
some qc files. Using these things you can quickly and reproducibly
generate a large number of plots. I will point out the points where you
want to save these objects in this vignette.</p>
<p>A further note is that I am in the practice of packaging all of the
pre-computed data objects into formal R packages. This aids portability,
prevents unintentional overwriting, lost data, manages
versions…basically a lot of good things for rigorous reproducible
science. However, building R packages is outside the scope of this
vignette. I may create a vignette on this in the future, but in the mean
time here is a very good source: <a href="https://kbroman.org/pkg_primer/" class="external-link uri">https://kbroman.org/pkg_primer/</a> This is where I learned
90% of how to build an R package.</p>
<p>Good luck!!</p>
</div>
</div>
<div class="section level2">
<h2 id="pre-processing-data-using-10x-genomics-cloud">Pre-processing data using 10X genomics cloud<a class="anchor" aria-label="anchor" href="#pre-processing-data-using-10x-genomics-cloud"></a>
</h2>
<p>As of 2021/2022, the preferred method of going from sequence data
(FASTQs) to cell-barcode matrices is using the 10X genomics cloud: <a href="https://www.10xgenomics.com/products/cloud-analysis" class="external-link uri">https://www.10xgenomics.com/products/cloud-analysis</a>. The
benefits over running this locally are too numerous to list. The only
drawbacks are that it requires some time to upload your data and you
have to do some point and click to get things configured. So anyways, if
you are working with FASTQs, this is what you should do.</p>
<p>Using 10X genomics cloud is free for the most part. It only restricts
the number of downloads you can perform for a given dataset. It does
require some very basic familiarity with the unix/linux command-line
interface, but it does provide premade commands for you to copy and
paste into the terminal.</p>
<p>The 10X Cloud outputs a set of files I refer to as a “pipestance”.
This is an archaic term dating from the time when we had to run the
analysis on our local machine. You will want to download the whole set
of files into their own directory with one subdirectory for each
biological sample. Including BAM files and dimensionality reduction in
the 10X Cloud analysis is optional and I recommend against it in most
cases. Unless you are doing things like RNA velocity or spliceoform
analysis you don’t need the BAMs and they are very large. We will be
doing our own dimensionality reduction and clustering so we don’t need
those files either. Once you have downloaded the files you should not
edit, move or rename them.</p>
<p>The recommended approach is to download the pipestance directly to
high-capacity storage, usually on an institutional network drive. That
way you don’t have to worry about backup or size. Once the data is
archived there you will read from it very few times so data transfer
time is not really an issue. Most of your downstream work will be with
compressed R objects which are small and portable. I usually keep the
FASTQs and the pipestances for each project together. You’ll want to
make it very clear using either README files or with the file structure
itself, which FASTQs generated which pipestances. It isn’t scripted (the
drawback I mentioned above) so if you want to know what you did a year
from now, you have to find a way to make it completely obvious to
yourself.</p>
</div>
<div class="section level2">
<h2 id="loading-data-into-a-list-of-celldataset-objects">Loading data into a list of cellDataSet objects<a class="anchor" aria-label="anchor" href="#loading-data-into-a-list-of-celldataset-objects"></a>
</h2>
<p>The first thing you have to do is load the 10X Data for each specimen
into a list of cellDataSet objects. You will use a configuration file to
do this. First some more detail on these topics:</p>
<ul>
<li>
<p>10X Data: The most important file is a zipped tar file containing
the cell barcodes, genes, and gene/barcode matrix. Depending on the
specific analysis pipeline run, it may be named
“filtered_feature_bc_matrix.tar.gz”, or
“sample_feature_bc_matrix.tar.gz”. The difference depends whether you
are running the “counts” pipeline or the “multi” pipeline. You don’t
want the raw_feature_bc_matrix which contains droplets below the UMI
cutoff (basically empty droplets) and un-de-multiplexed data if you are
using multiplexing.</p>
<p>The other things you will most likely want but we won’t cover in
detail here are the metrics summary file and any VDJ sequencing if you
are doing that. It is easier just to look at the html metrics summary
file but if you want to programmatically look at the sequencing metrics
file in R you can read that in. VDJ may be covered in another
vignette.</p>
</li>
<li><p>cellDataSet: Abbreviated CDS, this is the main data structure for
holding the single cell data. It is derived from the Bioconductor
SingleCellExperiment class. It can be thought of as a database holding a
table of cell metadata, a table of gene metadata, and matrix of
expression data, all of which are built from the pipestance data you
load in. There are additional “slots” available in the CDS class for
reduced dimensions of various types and cluster assignments. These we
don’t interact with directly for the most part. I find the data
structure very simple and straightforward to understand and work with.
All of the scRNA-seq functions in this package operate on this data
structure. I find the Seurat data structure less user-friendly. Where we
need to use Seurat-only functions (like label transfer and
Doubletfinder), Seurat objects are generated “behind the scenes”, so you
don’t really need to know them well to use the blaseRtools functions. If
you understood this paragraph then you already understand the CDS data
structure which is half the battle.</p></li>
<li><p>configuration file: Abbreviated config, this is a text file (csv)
that you can edit in excel and which will hold sample-level metadata,
including the file path where the pipestance is stored. You can see an
example I used for this vignette here:
<code>system.file("extdata/vignette_config.csv", package = "blaseRdata")</code>.
At a minimum it should have a column named “sample” and a column named
“targz_path” (the name for the latter is not critical).</p></li>
<li><p>sample: It is natural for most people to refer to the biological
specimen that produced the data as a “sample”. However this term gets
rather overloaded in these analyses, meaning that it gets used for too
many things in too many different places. If you want to be safer, you
can call it “specimen” or something else. If you want to use “sample” as
I do here, you have to make sure it is not added explicitly as a cell
metadata column. If you do it will cause an error with a function that
will try to create a “sample” column automatically. Hopefully that will
be more clear below.</p></li>
</ul>
<p>One final note: since some of the data processing are time-consuming,
most of the code listed here will not be executed with the vignette. You
can copy the code blocks, modify and use them on your own if you want. I
have included a final CDS object which you can access with
<code><a href="https://rdrr.io/pkg/blaseRdata/man/vignette_cds.html" class="external-link">blaseRdata::vignette_cds</a></code>. You can use this to make the
plots and do differential gene expression if you like.</p>
<p>Getting started then:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attach the packages you will need for the analysis.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/blaserlab/blaseRtools" class="external-link">blaseRtools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/blaserlab/blaseRdata" class="external-link">blaseRdata</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">monocle3</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://conflicted.r-lib.org/" class="external-link">conflicted</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://conflicted.r-lib.org/reference/conflict_prefer.html" class="external-link">conflict_prefer</a></span><span class="op">(</span><span class="st">"filter"</span>, <span class="st">"dplyr"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span>font_size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Read in the analysis configuration file. Here I am reading it from
the blaseRdata package but you would substitute your own file path.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in and inspect the configuration file.</span></span>
<span><span class="va">vignette_config</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/vignette_config.csv"</span>, </span>
<span>                                        package <span class="op">=</span> <span class="st">"blaseRdata"</span><span class="op">)</span>, </span>
<span>                            col_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>.default <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/parse_atomic.html" class="external-link">col_character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vignette_config</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 4</span></span></span>
<span><span class="co">#&gt;   sample              date     equipment targz_path                             </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chromium_controller 20211226 chromium  <span style="color: #949494;">"</span>X:\\Labs\\Blaser\\single_cell\\vignet…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chromium_X          20211226 X         <span style="color: #949494;">"</span>X:\\Labs\\Blaser\\single_cell\\vignet…</span></span></code></pre></div>
<p>The data we are using here is a very small sample dataset provided by
10X. It includes 2 PBMC samples run on different 10X machines: the
standard Chromium machine and the new Chromium X.</p>
<p>You’ll notice that I added a couple of sample metadata features in
addition to the required information. We will add those to the cds. Also
notice the format of the file path. This is linux reading a windows file
path. In the config file, I used the windows “copy path” feature and
simply pasted it into the csv. The double backslashes stand in for the
single backslash which is basically toxic to anything running on linux.
So this is not a problem. However, the X: stands for our network drive
where the data are located. Linux doesn’t know how to access that. So I
have a helper function that will translate it for you. This only works
with the OSUMC X drive. If you are using another drive you will have to
translate that to a linux-compatible file path manually. Here is how you
use that function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Fix the windows-style file path.</span></span>
<span><span class="va">vignette_config</span> <span class="op">&lt;-</span> <span class="va">vignette_config</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>targz_path <span class="op">=</span> <span class="fu"><a href="../reference/bb_fix_file_path.html">bb_fix_file_path</a></span><span class="op">(</span><span class="va">targz_path</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vignette_config</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 4</span></span></span>
<span><span class="co">#&gt;   sample              date     equipment targz_path                             </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chromium_controller 20211226 chromium  ~/network/X/Labs/Blaser/single_cell/vi…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chromium_X          20211226 X         ~/network/X/Labs/Blaser/single_cell/vi…</span></span></code></pre></div>
<p>This uses the pipe notation and tidyverse functions, which you should
become familiar with if you aren’t already.</p>
<p>Now we are ready to read in the data using the bb_load_tenx_targz
function. We will use the “apply” functional programming paradigm and
the purrr package to map this function across each sample in the config
file. This will produce a list of cds objects for us.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate a list of CDS objects using purrr::map</span></span>
<span><span class="va">cds_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span></span>
<span>  .x <span class="op">=</span> <span class="va">vignette_config</span><span class="op">$</span><span class="va">sample</span>,</span>
<span>  .f <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">conf</span> <span class="op">=</span> <span class="va">vignette_config</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">conf_filtered</span> <span class="op">&lt;-</span> <span class="va">conf</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample</span> <span class="op">==</span> <span class="va">x</span><span class="op">)</span></span>
<span>    <span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_load_tenx_targz.html">bb_load_tenx_targz</a></span><span class="op">(</span>targz_file <span class="op">=</span> <span class="va">conf_filtered</span><span class="op">$</span><span class="va">targz_path</span>,</span>
<span>                              sample_metadata_tbl <span class="op">=</span> <span class="va">conf_filtered</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                                <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">targz_path</span><span class="op">)</span><span class="op">)</span></span>
<span>                              <span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rlang.r-lib.org/reference/set_names.html" class="external-link">set_names</a></span><span class="op">(</span>nm <span class="op">=</span> <span class="va">vignette_config</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span></code></pre></div>
<p>Note that the line <code>select(-c(sample, targz_path))</code>
removes the sample column and the file path from the sample metadata. We
don’t need the file path. As mentioned above, having “sample” in the
cell metadata will cause an error later. We use
<code>set_names(nm = vignette_config$sample)</code> to name the CDS
object with the sample name. That will get added into the cell metadata
later.</p>
</div>
<div class="section level2">
<h2 id="qc-functions">QC Functions<a class="anchor" aria-label="anchor" href="#qc-functions"></a>
</h2>
<p>Next we want to remove low quality cells from the analysis. First we
identify low quality cells based on a high percentage of reads mapped to
mitochondrial genes or a low number of genes detected. We will map the
function bb_qc to each element of cds_list. bb_qc itself returns a list
of data objects, so we will get a list of lists. The qc calls will be
returned to the CDS objects later, but the whole output of this step is
worth saving for future reference.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># generate a list of qc results for individual CDS objects</span></span>
<span><span class="va">vig_qc_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/pmap.html" class="external-link">pmap</a></span><span class="op">(</span>.l <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>cds <span class="op">=</span> <span class="va">cds_list</span>,</span>
<span>                             cds_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cds_list</span><span class="op">)</span>,</span>
<span>                             genome <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"human"</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cds_list</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                   .f <span class="op">=</span> <span class="va">bb_qc</span></span>
<span>                   <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rlang.r-lib.org/reference/set_names.html" class="external-link">set_names</a></span><span class="op">(</span>nm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cds_list</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For example:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vig_qc_res</span><span class="op">$</span><span class="va">chromium_controller</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="va">vig_qc_res</span><span class="op">$</span><span class="va">chromium_controller</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span></span></code></pre></div>
<p>Next we want to remove potential cell doublets. We identify these
using a function from the <em>Doubletfinder</em> package. This generates
“pseudodoublets” and marks any real cells that map in the same area as
the pseudodoublets.</p>
<p>First we have to figure out the anticipated doublet rate which is
estimated by the number of cells in the cds/100000.</p>
<p>We also have to supply the qc results so we only run the prediction
on high-quality cells.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># gets the number of cells in each cds and divides it by 100000</span></span>
<span><span class="va">anticipated_doublet_rate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">cds_list</span>, <span class="va">ncol</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">100000</span></span>
<span></span>
<span><span class="co"># extracts the first element of the qc result list for each cds</span></span>
<span><span class="va">qc_calls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">vig_qc_res</span>,<span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generates a list of tables with doubletfinder results</span></span>
<span><span class="va">doubletfinder_list</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/pmap.html" class="external-link">pmap</a></span><span class="op">(</span></span>
<span>    .l <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      cds <span class="op">=</span> <span class="va">cds_list</span>,</span>
<span>      doublet_prediction <span class="op">=</span> <span class="va">anticipated_doublet_rate</span>,</span>
<span>      qc_table <span class="op">=</span> <span class="va">qc_calls</span></span>
<span>    <span class="op">)</span>,</span>
<span>    .f <span class="op">=</span> <span class="va">bb_doubletfinder</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rlang.r-lib.org/reference/set_names.html" class="external-link">set_names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cds_list</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now join the qc data and the doubletfinder data back onto
cds_list</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rejoins doubletfinder and qc data onto the list of CDS objects</span></span>
<span><span class="va">cds_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/pmap.html" class="external-link">pmap</a></span><span class="op">(</span></span>
<span>  .l <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    cds <span class="op">=</span> <span class="va">cds_list</span>,</span>
<span>    qc_data <span class="op">=</span> <span class="va">qc_calls</span>,</span>
<span>    doubletfinder_data <span class="op">=</span> <span class="va">doubletfinder_list</span></span>
<span>  <span class="op">)</span>,</span>
<span>  .f <span class="op">=</span> <span class="va">bb_rejoin</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="merging-into-a-single-celldataset-object-">Merging into a single cellDataSet object.<a class="anchor" aria-label="anchor" href="#merging-into-a-single-celldataset-object-"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Merge the CDS list into a single CDS</span></span>
<span><span class="va">vignette_cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/combine_cds.html" class="external-link">combine_cds</a></span><span class="op">(</span>cds_list <span class="op">=</span> <span class="va">cds_list</span><span class="op">)</span></span></code></pre></div>
<p>We can now remove some genes that are not going to be helpful. Since
we have already used the mitochondrial genes to identify poor-quality
cells, and since we are not interested in looking at them specifically
we can remove them. Likewise we can typically remove any ribosomal RNA
genes. Pre-calculated lists of genes to remove for human, mouse, and
zebrafish are provided in the blaseRdata package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove mitochondrial and ribosomal genes.</span></span>
<span><span class="va">vignette_cds</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">vignette_cds</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span><span class="op">$</span><span class="va">gene_short_name</span> <span class="op"><a href="../reference/grapes-notin-grapes.html">%notin%</a></span> <span class="va">hg38_remove_genes</span>,<span class="op">]</span></span></code></pre></div>
<p>Now we can remove the low quality cells and cell doublets. The
doubletfinder function provides doublet calls at two confidence levels.
If you like you can previzualize and select which ones you want to
remove. I ususally just remove the high confidence doublets.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove the low-quality cells</span></span>
<span><span class="va">vignette_cds</span> <span class="op">&lt;-</span> <span class="va">vignette_cds</span><span class="op">[</span>,<span class="fu">colData</span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span><span class="op">$</span><span class="va">qc.any</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove the high-confidence doublets</span></span>
<span><span class="va">vignette_cds</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">vignette_cds</span><span class="op">[</span>,<span class="fu">colData</span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span><span class="op">$</span><span class="va">doubletfinder_high_conf</span> <span class="op">==</span> <span class="st">"Singlet"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Now remove the qc and doubletfinder columns from the cell metadata because we are done with them.</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span><span class="op">$</span><span class="va">qc.any</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span><span class="op">$</span><span class="va">doubletfinder_low_conf</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span><span class="op">$</span><span class="va">doubletfinder_high_conf</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dimension-reduction">Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h2>
<p>Now we are ready to perform the dimensionality reduction. This is
done in two steps and we use unmodified monocle functions. First, we
calculate PCA components for each cell to reduce the number of
dimensions from the number of genes (thousands) to something like 50.
Monocle provides an option to set the number of PCAs you want to
calculate. Some people like exploring different PCAs and picking exactly
how many you want to calculate. I don’t find this to be a particularly
useful exercise so I usually just run it with the default setting which
is 50. The additional variance explained with PCAs 51 and greater is
tiny.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the PCA dimensions</span></span>
<span><span class="va">vignette_cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/preprocess_cds.html" class="external-link">preprocess_cds</a></span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span></span></code></pre></div>
<p>Now we can run the UMAP algorithm to reduce the dimensions down to 2.
There are several options available here. You should set the number of
cores the algorithm will use.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate UMAP dimensions</span></span>
<span><span class="va">vignette_cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/reduce_dimension.html" class="external-link">reduce_dimension</a></span><span class="op">(</span><span class="va">vignette_cds</span>, cores <span class="op">=</span> <span class="fl">40</span><span class="op">)</span></span></code></pre></div>
<p>Running UMAP with multiple cores will generate slightly different
results each time you run it, so you want to get to this point and then
save your data object with
<code>save(vignette_cds, file = "&lt;directory&gt;/vignette_cds.rda", compress = "gzip")</code>.</p>
<p>At this point it is worth exploring the cell and gene metadata to
understand how things are arranged. We can use the bb_cellmeta and
bb_rowmeta convenience functions to return the metadata in the form of a
tibble. This is more convenient for exploring and manipulating, but if
you want to modify or add back to the CDS object, you should use the
form <code>colData(CDS)$new_column &lt;- data</code></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cell metadata</span></span>
<span><span class="fu"><a href="../reference/bb_cellmeta.html">bb_cellmeta</a></span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span> </span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,160 × 16</span></span></span>
<span><span class="co">#&gt;    cell_id          barcode Size_Factor date  equipment sample prealignment_dim1</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> AATCACGAGGATCAC… AATCAC…       3.20  2021… chromium  chrom…              9.42</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> AATCACGAGTGGCCT… AATCAC…       0.990 2021… chromium  chrom…             -<span style="color: #BB0000;">7.10</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> AATCACGCACAGAGA… AATCAC…       0.751 2021… chromium  chrom…             -<span style="color: #BB0000;">4.14</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> AATCACGCACCAGCG… AATCAC…       3.08  2021… chromium  chrom…              6.97</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> AATCACGCAGGTTCA… AATCAC…       0.627 2021… chromium  chrom…             -<span style="color: #BB0000;">4.12</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> AATCACGCATATCTG… AATCAC…       0.830 2021… chromium  chrom…             -<span style="color: #BB0000;">3.56</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> AATCACGGTCAAGCG… AATCAC…       0.931 2021… chromium  chrom…             -<span style="color: #BB0000;">4.04</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> AATCACGGTTGGACC… AATCAC…       0.596 2021… chromium  chrom…             -<span style="color: #BB0000;">7.70</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> AATCACGTCCCGTTG… AATCAC…       1.16  2021… chromium  chrom…             -<span style="color: #BB0000;">3.28</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> AATCACGTCGTTCCT… AATCAC…       0.746 2021… chromium  chrom…             -<span style="color: #BB0000;">5.68</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,150 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 9 more variables: prealignment_dim2 &lt;dbl&gt;, leiden &lt;fct&gt;, partition &lt;fct&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   louvain &lt;fct&gt;, seurat_dim1 &lt;dbl&gt;, seurat_dim2 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   seurat_celltype_l1 &lt;chr&gt;, seurat_celltype_l2 &lt;chr&gt;, leiden_assignment &lt;fct&gt;</span></span></span>
<span></span>
<span><span class="co"># Gene metadata</span></span>
<span><span class="fu"><a href="../reference/bb_rowmeta.html">bb_rowmeta</a></span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 36,398 × 8</span></span></span>
<span><span class="co">#&gt;    feature_id  id    gene_short_name data_type module module_labeled supermodule</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ENSG000002… ENSG… MIR1302-2HG     Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ENSG000002… ENSG… FAM138A         Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ENSG000001… ENSG… OR4F5           Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ENSG000002… ENSG… AL627309.1      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ENSG000002… ENSG… AL627309.3      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ENSG000002… ENSG… AL627309.2      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ENSG000002… ENSG… AL627309.5      Gene Exp… 1      Module 1       1          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ENSG000002… ENSG… AL627309.4      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ENSG000002… ENSG… AP006222.2      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ENSG000002… ENSG… AL732372.1      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36,388 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: supermodule_labeled &lt;fct&gt;</span></span></span></code></pre></div>
<p>Note that the sample names have been added into the cell
metadata.</p>
<p>At this point you want to visualize your cells and determine if you
need to perform any batch-correction techniques.</p>
</div>
<div class="section level2">
<h2 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h2>
<p>The data from these samples is very similar but let’s say there was
some batch effect you wanted to reduce between these two samples. You
can align any number of samples by selecting a cell metadata column to
align by. This will reduce the variability in UMAP dimension based on
that column. If you want to align by more than one column, you should
create a composite variable of all columns you want to align by, say by
date and facility.</p>
<p>The important thing to remember here is that alignment/batch
correction does not change the underlying expression data. It only
changes the PCAs and UMAP calculations. So if you also want to account
for these variables when doing differential gene expression, then you
need to model that in your calculations (see below). So why do batch
correction? It helps reduce the heterogeneity of your dataset so you can
put similar cells together in clusters. This is an important concept for
finding differences between samples.</p>
<p>The function bb_align performs these steps: calculate aligned PCA
dimensions, recalculate umap dimensions, add the pre-alignment
dimensions into the cell metadata in case you want to use them
again.</p>
<p>Usually I will save this to a temporary object until I find a useful
alignment formula. Then we can save as the original CDS and discard the
temporary object once we are satisfied.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Align samples according to the equipment metadata column</span></span>
<span><span class="va">vignette_cds_aligned_temp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_align.html">bb_align</a></span><span class="op">(</span><span class="va">vignette_cds</span>, align_by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Replace the original CDS with the Aligned CDS</span></span>
<span><span class="va">vignette_cds</span> <span class="op">&lt;-</span> <span class="va">vignette_cds_aligned_temp</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">vignette_cds_aligned_temp</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, var <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-16-1.png" width="576" style="display: block; margin: auto;"></p>
<p>You can see that the pre-alignment dimensions are stored in the cell
metadata so you can plot them later if you want.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_cellmeta.html">bb_cellmeta</a></span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,160 × 16</span></span></span>
<span><span class="co">#&gt;    cell_id          barcode Size_Factor date  equipment sample prealignment_dim1</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> AATCACGAGGATCAC… AATCAC…       3.20  2021… chromium  chrom…              9.42</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> AATCACGAGTGGCCT… AATCAC…       0.990 2021… chromium  chrom…             -<span style="color: #BB0000;">7.10</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> AATCACGCACAGAGA… AATCAC…       0.751 2021… chromium  chrom…             -<span style="color: #BB0000;">4.14</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> AATCACGCACCAGCG… AATCAC…       3.08  2021… chromium  chrom…              6.97</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> AATCACGCAGGTTCA… AATCAC…       0.627 2021… chromium  chrom…             -<span style="color: #BB0000;">4.12</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> AATCACGCATATCTG… AATCAC…       0.830 2021… chromium  chrom…             -<span style="color: #BB0000;">3.56</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> AATCACGGTCAAGCG… AATCAC…       0.931 2021… chromium  chrom…             -<span style="color: #BB0000;">4.04</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> AATCACGGTTGGACC… AATCAC…       0.596 2021… chromium  chrom…             -<span style="color: #BB0000;">7.70</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> AATCACGTCCCGTTG… AATCAC…       1.16  2021… chromium  chrom…             -<span style="color: #BB0000;">3.28</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> AATCACGTCGTTCCT… AATCAC…       0.746 2021… chromium  chrom…             -<span style="color: #BB0000;">5.68</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,150 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 9 more variables: prealignment_dim2 &lt;dbl&gt;, leiden &lt;fct&gt;, partition &lt;fct&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   louvain &lt;fct&gt;, seurat_dim1 &lt;dbl&gt;, seurat_dim2 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   seurat_celltype_l1 &lt;chr&gt;, seurat_celltype_l2 &lt;chr&gt;, leiden_assignment &lt;fct&gt;</span></span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, var <span class="op">=</span> <span class="st">"sample"</span>, </span>
<span>            alt_dim_x <span class="op">=</span> <span class="st">"prealignment_dim1"</span>, </span>
<span>            alt_dim_y <span class="op">=</span> <span class="st">"prealignment_dim2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-18-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<p>There are three clustering methods we use. Louvain (ref: PMID
30914743) are the finest/smallest, Leiden (ref: PMID 30914743) are
medium, and Partitions (ref: PMID 30890159) are the largest/coarsest. We
have a function to calculate all three using default parameters and add
them explicitly into the CDS cell metadata. In addition, this function
will calculate the top markers for each and produce a csv file which you
can look at externally or read back into R as I do here. Usually you
want to read it back into R so when you identify what the clusters are,
you can label them in the table.</p>
<p>Using this function requires using default clustering parameters in
the underlying monocle functions. If you want to modify those (usually
not necessary) you need to use the original monocle functions.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify clusters and calculate top markers</span></span>
<span><span class="va">marker_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">vignette_cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_triplecluster.html">bb_triplecluster</a></span><span class="op">(</span><span class="va">vignette_cds</span>, n_top_markers <span class="op">=</span> <span class="fl">50</span>, outfile <span class="op">=</span> <span class="va">marker_file</span>, n_cores <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">vignette_top_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="va">marker_file</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vignette_top_markers</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,100 × 11</span></span></span>
<span><span class="co">#&gt;    gene_id         gene_short_name cluster_method cell_group  marker_score</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ENSG00000265972 TXNIP           partition      partition 1        0.998</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ENSG00000081237 PTPRC           partition      partition 1        0.997</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ENSG00000163041 H3F3A           partition      partition 1        0.997</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ENSG00000034510 TMSB10          partition      partition 1        1    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ENSG00000114942 EEF1B2          partition      partition 1        0.991</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ENSG00000187514 PTMA            partition      partition 1        1    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ENSG00000232112 TMA7            partition      partition 1        0.984</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ENSG00000269028 MTRNR2L12       partition      partition 1        0.998</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ENSG00000145741 BTF3            partition      partition 1        0.985</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ENSG00000127184 COX7C           partition      partition 1        0.991</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,090 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 6 more variables: mean_expression &lt;dbl&gt;, fraction_expressing &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   specificity &lt;dbl&gt;, pseudo_R2 &lt;dbl&gt;, marker_test_p_value &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   marker_test_q_value &lt;dbl&gt;</span></span></span></code></pre></div>
<p>Now you can plot the cells and color by cluster type</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, var <span class="op">=</span> <span class="st">"partition"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-21-1.png" width="432" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, var <span class="op">=</span> <span class="st">"leiden"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-21-2.png" width="432" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, var <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-21-3.png" width="432" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="gene-modules">Gene modules<a class="anchor" aria-label="anchor" href="#gene-modules"></a>
</h2>
<p>Gene modules are groups of genes that are similarly co-regulated from
cell to cell in the data set. We use monocle functions to identify gene
modules and store them as gene metadata.</p>
<p>Gene modules in this sense can only be defined using single cell
data. Essentially the algorithm reverses the UMAP calculation and plots
“coordinates” for each gene according to the cell it is expressed in.
Then it clusters the genes according to the UMAP coordinates just like
we did for the cells. These are modules and super modules.</p>
<p>Gene modules can have dozens or thousands of genes depending on your
dataset. If they are evenly expressed across all cells (like
housekeeping genes), some genes may not end up in a module.</p>
<p>Note: this is a computationally-intensive operation which may
overload your computer if you use too many cores. It is also
non-deterministic so you need to save the CDS object when you are
done.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify gene modules and add them to the gene metadata.</span></span>
<span><span class="va">vignette_cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_gene_modules.html">bb_gene_modules</a></span><span class="op">(</span><span class="va">vignette_cds</span>, n_cores <span class="op">=</span> <span class="fl">24</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_rowmeta.html">bb_rowmeta</a></span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 36,398 × 8</span></span></span>
<span><span class="co">#&gt;    feature_id  id    gene_short_name data_type module module_labeled supermodule</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ENSG000002… ENSG… MIR1302-2HG     Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ENSG000002… ENSG… FAM138A         Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ENSG000001… ENSG… OR4F5           Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ENSG000002… ENSG… AL627309.1      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ENSG000002… ENSG… AL627309.3      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ENSG000002… ENSG… AL627309.2      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ENSG000002… ENSG… AL627309.5      Gene Exp… 1      Module 1       1          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ENSG000002… ENSG… AL627309.4      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ENSG000002… ENSG… AP006222.2      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ENSG000002… ENSG… AL732372.1      Gene Exp… <span style="color: #BB0000;">NA</span>     No Module      <span style="color: #BB0000;">NA</span>         </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 36,388 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: supermodule_labeled &lt;fct&gt;</span></span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="label-transfer-using-seurat">Label transfer using Seurat<a class="anchor" aria-label="anchor" href="#label-transfer-using-seurat"></a>
</h2>
<p>This is an optional step you can use to generate automatic cell
labels. You need to take these results with a grain of salt and always
explore the automated cell labels before believing them.</p>
<p>For example if you are looking at leukemia cells and you generate
cell labels based on a normal PBMC dataset, you will provide normal
celltype labels to leukemia cells which is misleading.</p>
<p>Another issue is that there are many programs that claim to be able
to do this but only Seurat seems to be useful in my opinion. However the
Seurat people have put most of their effort in this area into developing
web apps which have significant limitations for us. Their raw reference
data is difficult to find and implement and is only available for human
PBMC. For now, we can only use this technique for human PBMC.</p>
<p>If you want to use this, you need to download this file: <a href="https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat" class="external-link uri">https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat</a>.
It is about 2GB so we don’t even distribute it with blaseRdata. You need
to download that and put it somewhere on your system where you are
running the analysis. Then run this, substituting in the appropriate
file path:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Annotate the PBMC data</span></span>
<span><span class="va">vignette_cds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_seurat_anno.html">bb_seurat_anno</a></span><span class="op">(</span><span class="va">vignette_cds</span>, reference <span class="op">=</span> <span class="st">"~/workspace_pipelines/sc_refdata/seurat_pbmc_reference_20210506/pbmc_multimodal.h5seurat"</span><span class="op">)</span></span></code></pre></div>
<p>This function has added two celltype assignments to the cell
metadata. It has also added suerat umap dims. These are where the cells
would’ve landed if they were processed in the reference dataset.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_cellmeta.html">bb_cellmeta</a></span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,160 × 16</span></span></span>
<span><span class="co">#&gt;    cell_id          barcode Size_Factor date  equipment sample prealignment_dim1</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> AATCACGAGGATCAC… AATCAC…       3.20  2021… chromium  chrom…              9.42</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> AATCACGAGTGGCCT… AATCAC…       0.990 2021… chromium  chrom…             -<span style="color: #BB0000;">7.10</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> AATCACGCACAGAGA… AATCAC…       0.751 2021… chromium  chrom…             -<span style="color: #BB0000;">4.14</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> AATCACGCACCAGCG… AATCAC…       3.08  2021… chromium  chrom…              6.97</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> AATCACGCAGGTTCA… AATCAC…       0.627 2021… chromium  chrom…             -<span style="color: #BB0000;">4.12</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> AATCACGCATATCTG… AATCAC…       0.830 2021… chromium  chrom…             -<span style="color: #BB0000;">3.56</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> AATCACGGTCAAGCG… AATCAC…       0.931 2021… chromium  chrom…             -<span style="color: #BB0000;">4.04</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> AATCACGGTTGGACC… AATCAC…       0.596 2021… chromium  chrom…             -<span style="color: #BB0000;">7.70</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> AATCACGTCCCGTTG… AATCAC…       1.16  2021… chromium  chrom…             -<span style="color: #BB0000;">3.28</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> AATCACGTCGTTCCT… AATCAC…       0.746 2021… chromium  chrom…             -<span style="color: #BB0000;">5.68</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1,150 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 9 more variables: prealignment_dim2 &lt;dbl&gt;, leiden &lt;fct&gt;, partition &lt;fct&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   louvain &lt;fct&gt;, seurat_dim1 &lt;dbl&gt;, seurat_dim2 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   seurat_celltype_l1 &lt;chr&gt;, seurat_celltype_l2 &lt;chr&gt;, leiden_assignment &lt;fct&gt;</span></span></span></code></pre></div>
<p>You can plot the cells using the Seurat coordinates:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, </span>
<span>            var <span class="op">=</span> <span class="st">"seurat_celltype_l1"</span>, </span>
<span>            alt_dim_x <span class="op">=</span> <span class="st">"seurat_dim1"</span>, </span>
<span>            alt_dim_y <span class="op">=</span> <span class="st">"seurat_dim2"</span>, </span>
<span>            overwrite_labels <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>            group_label_size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-26-1.png" width="528" style="display: block; margin: auto;"></p>
<p>Or you can plot them using their own dimension reduction coordinates.
I think this method is better because it reflects the data more
accurately.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, var <span class="op">=</span> <span class="st">"seurat_celltype_l1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-27-1.png" width="432" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="cell-assignments">Cell Assignments<a class="anchor" aria-label="anchor" href="#cell-assignments"></a>
</h2>
<p>This completes the basic loading and processing of single cell data
in terms of what can be scripted and run non-interactively. Usually what
I will do is build up a script of all of the preceeding commands and
then refresh the R session and source the whole script again. Even if it
takes hours to run again (doubletfinder, seurat annotations and gene
modules are computationally intensive), it is easy enough to let it run
overnight and then you can be confident that your output (the final CDS)
is really derived from the data and code you have in the script.</p>
<p>Once this is done, I make new scripts for other analysis steps, such
as Cell Assignments. Making cell assignments is scientifically
challenging but computationally very easy, so you want to put some
thought into it. I usually resave the CDS object with cell assignemnts
when I am done but these can be edited or new cell assignments can be
made very easily if you need to do so.</p>
<p>Usually you want to pick a cluster resolution that is going to be
useful and give all of the cells in that cluster the same name.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, var <span class="op">=</span> <span class="st">"leiden"</span>, plot_title <span class="op">=</span> <span class="st">"Leiden Clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-28-1.png" width="432" style="display: block; margin: auto;"></p>
<p>This is a low-throughput dataset with relatively low resolution, but
it illustrates the method.</p>
<p>There are a few resources you have available to make cell
assignments: 1. Cluster top markers. 2. Gene modules 3. Seurat
assignments. We already defined all of these. Gene modules are very
powerful for finding out what clusters are but it takes a few more steps
to extract those data which I will show later. Let’s say in this case
we’ve inspected the cluster top markers and we need some help from the
Seurat assignments. As I said previously, I never use the Seurat
assignments as is. You can explore them visually as above or you can do
something more quantitative.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leiden_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_cellmeta.html">bb_cellmeta</a></span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">leiden</span>, <span class="va">seurat_celltype_l1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">leiden_seurat</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   leiden [3]</span></span></span>
<span><span class="co">#&gt;   leiden seurat_celltype_l1     n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 1      CD4 T                307</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 1      CD8 T                207</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 1      NK                    48</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 1      other                  1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 1      other T               26</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 2      DC                    33</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> 2      Mono                 429</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> 3      B                    108</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> 3      other                  1</span></span></code></pre></div>
<p>That may be enough to tell you that Leiden cluster 1 is T/NK, Cluster
2 is Mono/DC and Cluster 3 is B cells.</p>
<p>If you have a complicated dataset you can use the more detailed
seurat_celltype_l2 assignments and/or you can plot it like this:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">leiden_seurat</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">leiden</span>, </span>
<span>                     y <span class="op">=</span> <span class="va">n</span>, </span>
<span>                     fill <span class="op">=</span> <span class="va">seurat_celltype_l1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>position<span class="op">=</span><span class="st">"fill"</span>, stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-30-1.png" width="432" style="display: block; margin: auto;"></p>
<p>Then we add a new cell metadata column by recoding the leiden column
into our designated assignments:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Recode the leiden clusters with our cell assignments</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span><span class="op">$</span><span class="va">leiden_assignment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html" class="external-link">recode</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span><span class="op">$</span><span class="va">leiden</span>, </span>
<span>                                                  <span class="st">"1"</span> <span class="op">=</span> <span class="st">"T/NK"</span>,</span>
<span>                                                  <span class="st">"2"</span> <span class="op">=</span> <span class="st">"DC/Mono"</span>,</span>
<span>                                                  <span class="st">"3"</span> <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, var <span class="op">=</span> <span class="st">"leiden_assignment"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-32-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="umap-plot-types">UMAP Plot Types<a class="anchor" aria-label="anchor" href="#umap-plot-types"></a>
</h2>
<p>We’ve already used the main umap plotting function, bb_var_umap. This
allows you to color the cells by any cell metadata variable you choose.
You set the variable name as the “var” argument. You can also highlight
a specific value.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, </span>
<span>            var <span class="op">=</span> <span class="st">"leiden_assignment"</span>, </span>
<span>            value_to_highlight <span class="op">=</span> <span class="st">"T/NK"</span>, </span>
<span>            cell_size <span class="op">=</span> <span class="fl">2</span>, </span>
<span>            foreground_alpha <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-33-1.png" width="432" style="display: block; margin: auto;"></p>
<p>This function returns a ggplot so if you don’t like the colors you
just have to add on a different scale. Color palettes can also be set
internally in the function.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, </span>
<span>            var <span class="op">=</span> <span class="st">"leiden_assignment"</span>, </span>
<span>            value_to_highlight <span class="op">=</span> <span class="st">"T/NK"</span>, </span>
<span>            palette <span class="op">=</span> <span class="st">"green4"</span>, </span>
<span>            cell_size <span class="op">=</span> <span class="fl">2</span>, </span>
<span>            foreground_alpha <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-34-1.png" width="432" style="display: block; margin: auto;"></p>
<p>An important concept for the overall analysis is understanding the
distribution of cell states between experimental conditions. We use a
density function to calculate the local density of cells in UMAP space
after faceting the plot by our experimental variable.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, </span>
<span>            var <span class="op">=</span> <span class="st">"density"</span>, </span>
<span>            facet_by <span class="op">=</span> <span class="st">"equipment"</span>, </span>
<span>            cell_size <span class="op">=</span> <span class="fl">2</span>, </span>
<span>            plot_title <span class="op">=</span> <span class="st">"Local Cell Density"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-35-1.png" width="672" style="display: block; margin: auto;"></p>
<p>As you can see, these are basically the same, which is what you would
expect.</p>
<p>Density is by definition normalized to the number of cells in the
local area. This can produce some over-normalization which may not be
useful. In this case you can map the color to the number of cells in
local bins. This may also work better with log normalization which you
can do. You would also want to use this with the option to downsample to
the same number of overall cells so your plot isn’t biased by the number
of cells recovered.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, </span>
<span>            var <span class="op">=</span> <span class="st">"local_n"</span>, </span>
<span>            nbin <span class="op">=</span> <span class="fl">10</span>, sample_equally <span class="op">=</span> <span class="cn">T</span>, </span>
<span>            facet_by <span class="op">=</span> <span class="st">"equipment"</span>, </span>
<span>            cell_size <span class="op">=</span> <span class="fl">2</span>, </span>
<span>            plot_title <span class="op">=</span> <span class="st">"Local N in 10 Bins"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-36-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/bb_var_umap.html">bb_var_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, </span>
<span>            var <span class="op">=</span> <span class="st">"log_local_n"</span>, </span>
<span>            nbin <span class="op">=</span> <span class="fl">10</span>, </span>
<span>            sample_equally <span class="op">=</span> <span class="cn">T</span>, </span>
<span>            facet_by <span class="op">=</span> <span class="st">"equipment"</span>, </span>
<span>            cell_size <span class="op">=</span> <span class="fl">2</span>, </span>
<span>            plot_title <span class="op">=</span> <span class="st">"Log Local N in 10 Bins"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-36-2.png" width="672" style="display: block; margin: auto;"></p>
<p>As in this case, usually Density looks better.</p>
<p>Since the concept of distribution between cell states is so
important, we have a dedicated function for plotting this.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_cluster_representation.html">bb_cluster_representation</a></span><span class="op">(</span>cds <span class="op">=</span> <span class="va">vignette_cds</span>, </span>
<span>                          cluster_var <span class="op">=</span> <span class="st">"leiden_assignment"</span>, </span>
<span>                          class_var <span class="op">=</span> <span class="st">"equipment"</span>, </span>
<span>                          experimental_class <span class="op">=</span> <span class="st">"X"</span>, </span>
<span>                          control_class <span class="op">=</span> <span class="st">"chromium"</span>, </span>
<span>                          return_value <span class="op">=</span> <span class="st">"plot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-37-1.png" width="432" style="display: block; margin: auto;"></p>
<p>This function tells you which class of sample is enriched in each
cluster. The values are normalized according to the number of total
cells captured for that class and presented as log fold change relative
to one of the classes (control) which you designate. Significance is
determined using Fisher’s exact test. You can test any number of
clusters but only two classes using this function. As shown above, this
function returns a ggplot which you can modify with any additional
ggplot layers you like. If this doesn’t work for you, you can ask the
function to return a data table which you can plot manually.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_cluster_representation.html">bb_cluster_representation</a></span><span class="op">(</span>cds <span class="op">=</span> <span class="va">vignette_cds</span>, </span>
<span>                          cluster_var <span class="op">=</span> <span class="st">"leiden_assignment"</span>, </span>
<span>                          class_var <span class="op">=</span> <span class="st">"equipment"</span>, </span>
<span>                          experimental_class <span class="op">=</span> <span class="st">"X"</span>, </span>
<span>                          control_class <span class="op">=</span> <span class="st">"chromium"</span>, </span>
<span>                          return_value <span class="op">=</span> <span class="st">"table"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 10</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   leiden_assignment [3]</span></span></span>
<span><span class="co">#&gt;   leiden_assignment     X chromium fold_change_over_con…¹ log2fold_change_over…²</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> T/NK              308.     285.                   1.08                 0.112  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> DC/Mono           232.     232.                   0.997               -<span style="color: #BB0000;">0.005</span><span style="color: #BB0000; text-decoration: underline;">01</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> B                  43.7     65.5                  0.667               -<span style="color: #BB0000;">0.583</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ abbreviated names: ¹​fold_change_over_control, ²​log2fold_change_over_control</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 5 more variables: enriched &lt;chr&gt;, n &lt;dbl&gt;, fisher_exact_p &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   p.signif &lt;chr&gt;, texty &lt;dbl&gt;</span></span></span></code></pre></div>
<p>This table shows the normalized cell count for each class of sample.
You can seee that after normalizing for the number of cells captured,
there were 64.5 B cells in the chromium sample and 42.6 B cells in the X
sample. This barely reaches signficance using the Fisher exact test. If
you look at the density and log local N UMAPs above, you can probably
appreciate the difference in cell color. I would say there is a subtle
difference but in the end it is up to your reader (or reviewer) if this
is convincing or not.</p>
<p>In addition to plotting cell metadata variables, we can plot gene
expression in UMAP plots.</p>
<p>For example:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_gene_umap.html">bb_gene_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, </span>
<span>             gene_or_genes <span class="op">=</span> <span class="st">"CD3D"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-39-1.png" width="432" style="display: block; margin: auto;"></p>
<p>Now a word of caution/advice. I find that these plots are most useful
for exploratory analysis and possibly supplemental figures but often not
main figures. The reason is that they convey less information per unit
area than other plot types, such as gene dotplots. It is often also more
difficult to visually distinguish differences between plots. This is
related mostly to the problem of overplotting. If you have many cells,
often the points will be plotted more or less on top of each other. If
you have a gene which is expressed in only a few of these cells, it will
be buried underneath all of the non-expressing cells. In order to
highlight these rare cells, this function makes the choice to plot all
non-expressing cells first on the bottom layer and then plot all of the
expressing cells on top. This ensures you can see the rare expressing
cells but may lead to overestimation of their number in some cases. So
it should be used with caution.</p>
<p>We can also plot multiple individual genes or aggregate gene scores
(e.g. modules) using this function:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_gene_umap.html">bb_gene_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>, gene_or_genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD19"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD14"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-40-1.png" width="720" style="display: block; margin: auto;"></p>
<p>Note that for plotting aggregate genes, the parameter gene_or_genes
should be provided with a data frame of two columns with the first
column being ensembl cell id and the second being the gene grouping.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">agg_genes</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/bb_rowmeta.html">bb_rowmeta</a></span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">id</span>, <span class="va">module_labeled</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">module_labeled</span> <span class="op">==</span> <span class="st">"Module 1"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/bb_gene_umap.html">bb_gene_umap</a></span><span class="op">(</span><span class="va">vignette_cds</span>,</span>
<span>             gene_or_genes <span class="op">=</span> <span class="va">agg_genes</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-41-1.png" width="432" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="gene-dotplots">Gene Dotplots<a class="anchor" aria-label="anchor" href="#gene-dotplots"></a>
</h2>
<p>A more information-dense way to plot these data is as a gene dotplot,
or “gene bubbles”. This plot is less susceptible to misleading the
reader because of overplotting.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_genebubbles.html">bb_genebubbles</a></span><span class="op">(</span><span class="va">vignette_cds</span>,</span>
<span>               genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"CD19"</span><span class="op">)</span>, </span>
<span>               cell_grouping <span class="op">=</span> <span class="st">"leiden_assignment"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-42-1.png" width="432" style="display: block; margin: auto;"></p>
<p>If run with default settings, this will order the genes and cell
groups using a biclustering algorithm. This is not apparent on the plot
above because of the small number of groups and genes. But you can also
specify the order:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_genebubbles.html">bb_genebubbles</a></span><span class="op">(</span><span class="va">vignette_cds</span>, </span>
<span>                genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"CD19"</span><span class="op">)</span>, </span>
<span>                cell_grouping <span class="op">=</span> <span class="st">"leiden_assignment"</span>, </span>
<span>                gene_ordering <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"as_supplied"</span><span class="op">)</span>,</span>
<span>                group_ordering <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"as_supplied"</span><span class="op">)</span><span class="op">)</span><span class="co">#will be alphabetical or by factor level</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-43-1.png" width="432" style="display: block; margin: auto;"></p>
<p>You can generate composite variables like so:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_genebubbles.html">bb_genebubbles</a></span><span class="op">(</span><span class="va">vignette_cds</span>, </span>
<span>                genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"CD19"</span><span class="op">)</span>, </span>
<span>                cell_grouping <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leiden_assignment"</span>, <span class="st">"louvain"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="scRNAseq_files/figure-html/unnamed-chunk-44-1.png" width="720" style="display: block; margin: auto;"></p>
<p>For more complicated figures, you can return a tibble and make the
plot on your own:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bb_genebubbles.html">bb_genebubbles</a></span><span class="op">(</span><span class="va">vignette_cds</span>, </span>
<span>               genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"CD19"</span><span class="op">)</span>, </span>
<span>               cell_grouping <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leiden_assignment"</span>, <span class="st">"louvain"</span><span class="op">)</span>,</span>
<span>               return_value <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 54 × 13</span></span></span>
<span><span class="co">#&gt;    group feature_id expression proportion id    gene_short_name data_type module</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> B_10  ENSG00000…    -<span style="color: #BB0000;">0.806</span>      0.090<span style="text-decoration: underline;">9</span> ENSG… CD14            Gene Exp… 1     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> B_10  ENSG00000…    -<span style="color: #BB0000;">0.669</span>      0.127  ENSG… CD3E            Gene Exp… 2     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> B_10  ENSG00000…     2.38       0.436  ENSG… CD19            Gene Exp… 3     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> B_14  ENSG00000…    -<span style="color: #BB0000;">0.804</span>      0.074<span style="text-decoration: underline;">1</span> ENSG… CD14            Gene Exp… 1     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> B_14  ENSG00000…    -<span style="color: #BB0000;">0.699</span>      0.074<span style="text-decoration: underline;">1</span> ENSG… CD3E            Gene Exp… 2     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> B_14  ENSG00000…     3.08       0.463  ENSG… CD19            Gene Exp… 3     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> DC/M… ENSG00000…    -<span style="color: #BB0000;">0.087</span><span style="color: #BB0000; text-decoration: underline;">6</span>     0.619  ENSG… CD14            Gene Exp… 1     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> DC/M… ENSG00000…    -<span style="color: #BB0000;">0.764</span>      0.047<span style="text-decoration: underline;">6</span> ENSG… CD3E            Gene Exp… 2     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> DC/M… ENSG00000…    -<span style="color: #BB0000;">0.355</span>      0      ENSG… CD19            Gene Exp… 3     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> DC/M… ENSG00000…     1.51       1      ENSG… CD14            Gene Exp… 1     </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 44 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 5 more variables: module_labeled &lt;fct&gt;, supermodule &lt;fct&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   supermodule_labeled &lt;fct&gt;, leiden_assignment &lt;fct&gt;, louvain &lt;fct&gt;</span></span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-gene-expression">Differential Gene Expression<a class="anchor" aria-label="anchor" href="#differential-gene-expression"></a>
</h2>
<p>At this point I think our example dataset demonstrates that outside
of a small difference in the number of B cells captured, there is a high
degree of similarity between the samples run on the chromium and the
chromium_X machines. Therefore it is plausible that we can use these as
experimental replicates to determine differential gene expression
between cell clusters.</p>
<p>This is an important point for your experimental design. Here we are
using these sample datasets because they are small and happen to be
available. This of course is not how you do experiemnts. You want to
design your experiment to include the best possible biological
replicates you can find: clutchmates or littermates if possible. If you
are working with patient samples you are not going to be able to achieve
this so you may want to try to model the potential confounding variables
like age, sex, treatments etc. in your analysis. Likewise if your mice
are coming from different litters or were processed in different batches
you may want to model those confounders. We have functions that can
handle this sort of complexity. The most robust of these is called
<em>pseudobulk analysis</em>. It extracts the counts from groups of
cells, generates an aggregate count profile and performs differential
gene expression according to the experimental model you provide.
Alternatively you can use a function based on <em>generalized linear
models</em> and <em>multivariate regression</em>. The latter function
does not require experimental replicates but has some drawbacks in terms
of inflated significance values.</p>
<div class="section level3">
<h3 id="pseudobulk-analysis">Pseudobulk Analysis<a class="anchor" aria-label="anchor" href="#pseudobulk-analysis"></a>
</h3>
<p>Pseudobulk analysis is generally considered more robust than other
methods of differential gene expression because it treats biological
samples as biological replciates rather than individual cells as
biological replicates.</p>
<p>In order to perform this, we aggregate the raw counts from groups of
cells that we have to define and then use a bulk RNAseq differential
expression program called DESeq2 to generate log fold change and p/q
values between the cell groups.</p>
<p>Setting this up requires a little advance thought. You will need to
consider that there may be fixed sample-level covariates that could
confound interpretation of the comparison you want to perform. For
example, let’s say you want to compare gene expression between a
heterogeneous cluster of cells from 20 patient samples; 10 patients are
responders and 10 are non-responders to therapy. Your goal is to
identify a set of genes that are predicted to be differentially
expressed based on the response variable (responders/non-responders).
However, you may know other things about these patients such as age,
sex, tumor histology, genotype etc. that also explain these differences
in gene expression or perhaps explain them better than response to
therapy. So you have to model these variables if you want to understand
this.</p>
<p><strong>If you are working with experimental animal models you should
never allow such confounders to enter your experiment. Always use true
biological replicates so that you may interpret differential gene
expression based on a single experimental variable.</strong></p>
<p>The next important feature to consider is that for your primary
comparison (say response to therapy) you must always have biological
replicates of some sort. In other words if you want to identify an
effect of response on gene expression, you must have a minimum of 2
responders and 2 non-responders.</p>
<p>You will need to come up with a table describing all of the variables
you may wish include in your experimental design. The best place to
start is with the CDS cell metadata. You will need the unique biological
replicate identifier, the variable you wish to compare and any
covariates that you wish to include in the design. In the example CDS we
are working with, “sample” identifies the biological replicates and
“leiden_assignment” identifies the cell groups we wish to compare. There
are no other non-redundant sample variables.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vignette_exp_design</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/bb_cellmeta.html">bb_cellmeta</a></span><span class="op">(</span><span class="va">vignette_cds</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">leiden_assignment</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">vignette_exp_design</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   sample [2]</span></span></span>
<span><span class="co">#&gt;   sample              leiden_assignment</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chromium_X          T/NK             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chromium_X          DC/Mono          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chromium_X          B                </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chromium_controller T/NK             </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chromium_controller DC/Mono          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chromium_controller B</span></span></code></pre></div>
<p>Each row in this table we will call a “pseudosample”. This is because
it represents a group of cells from either a biologically distinct
sample or cell cluster. For this simple comparison, we are now ready to
run the pseudobulk function.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vignette_pseudobulk_res</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/bb_pseudobulk_mf.html">bb_pseudobulk_mf</a></span><span class="op">(</span>cds <span class="op">=</span> <span class="va">vignette_cds</span>,</span>
<span>                   pseudosample_table <span class="op">=</span> <span class="va">vignette_exp_design</span>, </span>
<span>                   design_formula <span class="op">=</span> <span class="st">"~ leiden_assignment"</span>,</span>
<span>                   result_recipe <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leiden_assignment"</span>, <span class="st">"T/NK"</span>, <span class="st">"DC/Mono"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now explore the result:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Detailed column headers for the results tables.</span></span>
<span><span class="va">vignette_pseudobulk_res</span><span class="op">$</span><span class="va">Header</span> </span>
<span><span class="co">#&gt; [1] "mean of normalized counts for all samples"                </span></span>
<span><span class="co">#&gt; [2] "log2 fold change (MLE): leiden_assignment T_NK vs DC_Mono"</span></span>
<span><span class="co">#&gt; [3] "standard error: leiden_assignment T_NK vs DC_Mono"        </span></span>
<span><span class="co">#&gt; [4] "Wald statistic: leiden_assignment T_NK vs DC_Mono"        </span></span>
<span><span class="co">#&gt; [5] "Wald test p-value: leiden_assignment T_NK vs DC_Mono"     </span></span>
<span><span class="co">#&gt; [6] "BH adjusted p-values"</span></span></code></pre></div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Differential expression results.  Positive L2FC indicates up in T/NK vs DC/Mono</span></span>
<span></span>
<span><span class="va">vignette_pseudobulk_res</span><span class="op">$</span><span class="va">Result</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6,867 × 8</span></span></span>
<span><span class="co">#&gt;    id             gene_short_name baseMean log2FoldChange  lfcSE  stat    pvalue</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ENSG000000085… IL32                727.           6.08 0.138   44.0 0   <span style="color: #949494;"> </span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ENSG000001672… CD3D                418.           5.81 0.166   35.0 1.23<span style="color: #949494;">e</span><span style="color: #BB0000;">-268</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ENSG000001149… EEF1B2             <span style="text-decoration: underline;">2</span>433.           1.81 0.057<span style="text-decoration: underline;">8</span>  31.4 3.87<span style="color: #949494;">e</span><span style="color: #BB0000;">-216</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ENSG000002477… PCED1B-AS1          320.           4.44 0.146   30.4 2.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-203</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ENSG000002117… TRBC2               357.           5.67 0.187   30.3 2.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-201</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ENSG000002777… TRAC                276.           5.63 0.187   30.2 5.51<span style="color: #949494;">e</span><span style="color: #BB0000;">-200</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ENSG000001271… BCL11B              277.           5.97 0.199   30.1 1.13<span style="color: #949494;">e</span><span style="color: #BB0000;">-198</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ENSG000001117… LDHB                368.           2.91 0.098<span style="text-decoration: underline;">2</span>  29.6 2.80<span style="color: #949494;">e</span><span style="color: #BB0000;">-192</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ENSG000001988… CD3E                290.           6.02 0.204   29.6 6.28<span style="color: #949494;">e</span><span style="color: #BB0000;">-192</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ENSG000002715… CCL5                295.           5.44 0.188   28.8 6.53<span style="color: #949494;">e</span><span style="color: #BB0000;">-183</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 6,857 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: padj &lt;dbl&gt;</span></span></span>
<span>  </span>
<span><span class="va">vignette_pseudobulk_res</span><span class="op">$</span><span class="va">Result</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">log2FoldChange</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6,967 × 8</span></span></span>
<span><span class="co">#&gt;    id          gene_short_name baseMean log2FoldChange  lfcSE  stat pvalue  padj</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ENSG000001… CTSS               <span style="text-decoration: underline;">2</span>502.          -<span style="color: #BB0000;">5.11</span> 0.094<span style="text-decoration: underline;">0</span> -<span style="color: #BB0000;">54.3</span>      0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ENSG000001… S100A9             <span style="text-decoration: underline;">6</span>297.          -<span style="color: #BB0000;">5.80</span> 0.083<span style="text-decoration: underline;">5</span> -<span style="color: #BB0000;">69.5</span>      0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ENSG000001… S100A8             <span style="text-decoration: underline;">3</span>384.          -<span style="color: #BB0000;">5.82</span> 0.115  -<span style="color: #BB0000;">50.6</span>      0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ENSG000000… CD74               <span style="text-decoration: underline;">3</span>478.          -<span style="color: #BB0000;">4.23</span> 0.075<span style="text-decoration: underline;">9</span> -<span style="color: #BB0000;">55.7</span>      0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ENSG000002… LST1                633.          -<span style="color: #BB0000;">4.48</span> 0.116  -<span style="color: #BB0000;">38.6</span>      0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ENSG000002… AIF1               <span style="text-decoration: underline;">1</span>030.          -<span style="color: #BB0000;">4.36</span> 0.098<span style="text-decoration: underline;">2</span> -<span style="color: #BB0000;">44.4</span>      0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ENSG000002… HLA-DRA            <span style="text-decoration: underline;">3</span>382.          -<span style="color: #BB0000;">5.48</span> 0.109  -<span style="color: #BB0000;">50.2</span>      0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ENSG000002… HLA-DPA1           <span style="text-decoration: underline;">1</span>101.          -<span style="color: #BB0000;">4.83</span> 0.128  -<span style="color: #BB0000;">37.7</span>      0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ENSG000001… PSAP                990.          -<span style="color: #BB0000;">4.39</span> 0.097<span style="text-decoration: underline;">0</span> -<span style="color: #BB0000;">45.3</span>      0     0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ENSG000000… LYZ                <span style="text-decoration: underline;">5</span>176.          -<span style="color: #BB0000;">6.44</span> 0.132  -<span style="color: #BB0000;">48.8</span>      0     0</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 6,957 more rows</span></span></span></code></pre></div>
<p>These genes look as expected. These tables show pairwise comparisons.
Since there were three levels to the “leiden_assignment” variable, we
can select 2 of the three to compare in our “result_recipe”. The first
one lised will always be the “experimental” value and genes specific to
this condition will be positive while the second will be the “reference”
value and genes specific to this condition will be negative.</p>
<p>If you do not specify a “result_recipe” the program will revert to
default behavior which is to calculate results based on the final
variable in the design_formula using the first two values by
alphabetical order. This may or may not be what you want, so be aware of
this behavior.</p>
<p>This approach of making an experimental design based on the full CDS
and then selecting the result recipe you want to produce (technically
called “contrast”) is preferred rather than filtering the CDS to include
only T/NK and DC/Mono and running with default settings. This is because
the results vary somewhat and using the full CDS in this case gives a
more complete picture.</p>
<p>However, consider the case (which we will discuss but not demonstrate
here) where you have a sample-level variable of interest such as
response to therapy and you only want to look at the effect of this
variable on gene expression in a subset of cells. In this case it is OK
to subset the CDS to include only this subset of cells (say blasts or
leukemic cells) and then use a design formula such as “~ age + sex +
genotype + response”. This will give you the effect of response only in
blast cells, accounting for age, sex and genotype as potential
confounders.</p>
</div>
<div class="section level3">
<h3 id="multivariate-regression">Multivariate Regression<a class="anchor" aria-label="anchor" href="#multivariate-regression"></a>
</h3>
<p>This is a type of analysis you would want to use if you do not have
biological replicates suitable for the type of comparison you want to
make. In a way it considers each cell as a biological replicate which is
an arguable assumption many times.</p>
<p>The major practical limitations of this function are that it can
consider only a few genes at a time and does not generate a “log fold
change” that we would normally expect. Instead it generates an
“estimate” which is basically the slope of a regression line with p and
q values.</p>
<p>The regression function is based on the generalized linear regression
in R and by default uses the negative binomial distribution which is
similar to DESeq2 which we ahve already used.</p>
<p>The blaseRtools package has a function that wraps the multivariate
regression function from <em>monocle3</em> and add some minor
functionality.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vignette_regression_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bb_monocle_regression_better.html">bb_monocle_regression_better</a></span><span class="op">(</span>cds <span class="op">=</span> <span class="va">vignette_cds</span>, </span>
<span>                      gene_or_genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD19"</span>, <span class="st">"CD3D"</span>, <span class="st">"CD14"</span><span class="op">)</span>, </span>
<span>                      form <span class="op">=</span> <span class="st">"~leiden_assignment"</span><span class="op">)</span></span>
<span><span class="va">vignette_regression_res</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 8</span></span></span>
<span><span class="co">#&gt;   id       gene_short_name stratification formula term  log2FoldChange   p_value</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ENSG000… CD14            no stratifica… ~leide… leid…          6.59  1.28<span style="color: #949494;">e</span><span style="color: #BB0000;">-148</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ENSG000… CD14            no stratifica… ~leide… leid…          0.182 7.07<span style="color: #949494;">e</span><span style="color: #BB0000;">-  1</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ENSG000… CD14            no stratifica… ~leide… stat…          0.342 1.41<span style="color: #949494;">e</span><span style="color: #BB0000;">-  2</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ENSG000… CD3D            no stratifica… ~leide… leid…         -<span style="color: #BB0000;">5.87</span>  8.89<span style="color: #949494;">e</span><span style="color: #BB0000;">-237</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ENSG000… CD3D            no stratifica… ~leide… leid…         -<span style="color: #BB0000;">5.24</span>  4.44<span style="color: #949494;">e</span><span style="color: #BB0000;">- 49</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ENSG000… CD3D            no stratifica… ~leide… stat…          1.84  3.37<span style="color: #949494;">e</span><span style="color: #BB0000;">- 46</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> ENSG000… CD19            no stratifica… ~leide… leid…          0.174 6.18<span style="color: #949494;">e</span><span style="color: #BB0000;">-  1</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> ENSG000… CD19            no stratifica… ~leide… leid…          5.93  3.46<span style="color: #949494;">e</span><span style="color: #BB0000;">-  9</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> ENSG000… CD19            no stratifica… ~leide… stat…          0.360 4.37<span style="color: #949494;">e</span><span style="color: #BB0000;">-  3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: q_value &lt;dbl&gt;</span></span></span></code></pre></div>
<p>If you look at the term column you can see the comparisons being
made:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vignette_regression_res</span><span class="op">$</span><span class="va">term</span></span>
<span><span class="co">#&gt; [1] "leiden_assignmentDC/Mono"        "leiden_assignmentB"             </span></span>
<span><span class="co">#&gt; [3] "stats::offset(log(Size_Factor))" "leiden_assignmentDC/Mono"       </span></span>
<span><span class="co">#&gt; [5] "leiden_assignmentB"              "stats::offset(log(Size_Factor))"</span></span>
<span><span class="co">#&gt; [7] "leiden_assignmentDC/Mono"        "leiden_assignmentB"             </span></span>
<span><span class="co">#&gt; [9] "stats::offset(log(Size_Factor))"</span></span></code></pre></div>
<p>Here “leiden_assignmentDC/Mono” means that the comparison is DC/Mono
vs T/NK cells. You know this because T/NK does not show up in the term
column. It was chosen alphabetically as the reference condition. In this
case, a positive estimate means that the gene is up in DC/Mono vs T/NK
cells.</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Using these steps you should be able to go from sequencing data to
figures, gene lists and p values for many 10X datasets you come
across.</p>
<p>Please contact me by email with questions, corrections or
enhancements or post in github if you are so inclined.</p>
<p><a href="mailto:bradley.blaser@osumc.edu" class="email">bradley.blaser@osumc.edu</a> <a href="https://github.com/blaserlab/blaseRtools" class="external-link uri">https://github.com/blaserlab/blaseRtools</a></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
