<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Aggregate Single Cell Gene Expression — bb_aggregate • blaseRtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Aggregate Single Cell Gene Expression — bb_aggregate"><meta name="description" content="Generates a matrix of counts aggregated by gene and/or cell group."><meta property="og:description" content="Generates a matrix of counts aggregated by gene and/or cell group."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9199</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a></li>
    <li><a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a></li>
    <li><a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SummarizedHeatmap.html">Summarized Heatmap Vignette</a></li>
    <li><a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Aggregate Single Cell Gene Expression</h1>
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/R/bb_gene_umap.R" class="external-link"><code>R/bb_gene_umap.R</code></a></small>
      <div class="d-none name"><code>bb_aggregate.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generates a matrix of counts aggregated by gene and/or cell group.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bb_aggregate</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  experiment_type <span class="op">=</span> <span class="st">"Gene Expression"</span>,</span>
<span>  gene_group_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cell_group_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  norm_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"log"</span>, <span class="st">"binary"</span>, <span class="st">"size_only"</span><span class="op">)</span>,</span>
<span>  pseudocount <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  scale_agg_values <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  max_agg_value <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  min_agg_value <span class="op">=</span> <span class="op">-</span><span class="fl">3</span>,</span>
<span>  binary_min <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  exclude.na <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>A Seurat or cell data set object</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>Gene expression assay to use for aggregation; currently only applies to Seurat objects, Default: 'RNA'</p></dd>


<dt id="arg-gene-group-df">gene_group_df<a class="anchor" aria-label="anchor" href="#arg-gene-group-df"></a></dt>
<dd><p>A 2-column dataframe with gene names or ids and gene groupings, Default: NULL</p></dd>


<dt id="arg-cell-group-df">cell_group_df<a class="anchor" aria-label="anchor" href="#arg-cell-group-df"></a></dt>
<dd><p>A 2-coumn dataframe with cell ids and gene groupings, Default: NULL</p></dd>


<dt id="arg-norm-method">norm_method<a class="anchor" aria-label="anchor" href="#arg-norm-method"></a></dt>
<dd><p>Gene normalization method, Default: c("log", "binary", "size_only")</p></dd>


<dt id="arg-pseudocount">pseudocount<a class="anchor" aria-label="anchor" href="#arg-pseudocount"></a></dt>
<dd><p>Pseudocount, Default: 1</p></dd>


<dt id="arg-scale-agg-values">scale_agg_values<a class="anchor" aria-label="anchor" href="#arg-scale-agg-values"></a></dt>
<dd><p>Whether to scale the aggregated values, Default: TRUE</p></dd>


<dt id="arg-max-agg-value">max_agg_value<a class="anchor" aria-label="anchor" href="#arg-max-agg-value"></a></dt>
<dd><p>If scaling, make this the maximum aggregated value, Default: 3</p></dd>


<dt id="arg-min-agg-value">min_agg_value<a class="anchor" aria-label="anchor" href="#arg-min-agg-value"></a></dt>
<dd><p>If scaling, make this the minimum aggregated value, Default: -3</p></dd>


<dt id="arg-binary-min">binary_min<a class="anchor" aria-label="anchor" href="#arg-binary-min"></a></dt>
<dd><p>Minimum value below which a cell is considered not to express a feature, Default: 0</p></dd>


<dt id="arg-exclude-na">exclude.na<a class="anchor" aria-label="anchor" href="#arg-exclude-na"></a></dt>
<dd><p>Exclude NA?, Default: TRUE</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A dense or sparse matrix.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The best way to group genes or cells is by using bb_*meta and then select cell_id or feature_id plus one metadata column with your group labels.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://cli.r-lib.org/reference/cli_div.html" class="external-link">cli_div</a></code>, <code><a href="https://cli.r-lib.org/reference/cli_alert.html" class="external-link">cli_alert</a></code>
<code><a href="https://rdrr.io/pkg/monocle3/man/normalized_counts.html" class="external-link">normalized_counts</a></code>, <code>my.aggregate.Matrix</code>
<code>character(0)</code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

