<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Learn Graph and Calculate Pseudotime — bb_pseudotime • blaseRtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Learn Graph and Calculate Pseudotime — bb_pseudotime"><meta name="description" content="This function determines a gene expression trajectory using learn_graph from monocle3 and then calculates pseudotime dimensions along this trajectory using order_cells.  So it is 2 functions wrapped into 1.  Usually we will not adjust the parameters for learn_graph with the possible exception of close_loop and use_partition which are also available in this function with the same defaults.  If you need to fine-tune the trajectory, use monocle3::learn_graph on the cds object first and then run this function to calculate pseudotime.  The graph learning will not be repeated on an object unless force_graph is set to TRUE.
If you just want to look at the trajectory graph and not calculate pseudotime, change calculate_pseudotime to FALSE, or run monocle3::learn_graph.
After the pseudotime values are calculated, they are handled differently than in monocle3.  In this function, they are copied from the hidden CDS slot and made an explicit cell metadata column.  Pseudotime needs a starting point or anchor.  There is no interactive option here as in monocle3.  To identify this starting point, you identify a cell metadata variable and provide it to cluster_variable.  This should identify a cohesive group of cells in UMAP space such as a leiden cluster, louvain cluster or partition.  Then provide a value corresponding to the cluster of interest to cluster_value.  The function will start pseudotime at the cell closest to the graph node in that cluster.  The pseudotime value column will be named automatically as a composite of the cluster_variable and cluster_value parameters."><meta property="og:description" content="This function determines a gene expression trajectory using learn_graph from monocle3 and then calculates pseudotime dimensions along this trajectory using order_cells.  So it is 2 functions wrapped into 1.  Usually we will not adjust the parameters for learn_graph with the possible exception of close_loop and use_partition which are also available in this function with the same defaults.  If you need to fine-tune the trajectory, use monocle3::learn_graph on the cds object first and then run this function to calculate pseudotime.  The graph learning will not be repeated on an object unless force_graph is set to TRUE.
If you just want to look at the trajectory graph and not calculate pseudotime, change calculate_pseudotime to FALSE, or run monocle3::learn_graph.
After the pseudotime values are calculated, they are handled differently than in monocle3.  In this function, they are copied from the hidden CDS slot and made an explicit cell metadata column.  Pseudotime needs a starting point or anchor.  There is no interactive option here as in monocle3.  To identify this starting point, you identify a cell metadata variable and provide it to cluster_variable.  This should identify a cohesive group of cells in UMAP space such as a leiden cluster, louvain cluster or partition.  Then provide a value corresponding to the cluster of interest to cluster_value.  The function will start pseudotime at the cell closest to the graph node in that cluster.  The pseudotime value column will be named automatically as a composite of the cluster_variable and cluster_value parameters."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">blaseRtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9199</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Ape.html">Reading, Editing and Writing DNA Sequences with Ape</a></li>
    <li><a class="dropdown-item" href="../articles/blind_images.html">Blinding Images</a></li>
    <li><a class="dropdown-item" href="../articles/scRNAseq.html">scRNA-seq Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/SummarizedHeatmap.html">Summarized Heatmap Vignette</a></li>
    <li><a class="dropdown-item" href="../articles/trace.html">Trace:  Representing Track-like Data in R</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/blaserlab/blaseRtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Learn Graph and Calculate Pseudotime</h1>
      <small class="dont-index">Source: <a href="https://github.com/blaserlab/blaseRtools/blob/HEAD/R/bb_pseudotime.R" class="external-link"><code>R/bb_pseudotime.R</code></a></small>
      <div class="d-none name"><code>bb_pseudotime.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function determines a gene expression trajectory using <code>learn_graph</code> from monocle3 and then calculates pseudotime dimensions along this trajectory using <code>order_cells</code>.  So it is 2 functions wrapped into 1.  Usually we will not adjust the parameters for learn_graph with the possible exception of <code>close_loop</code> and <code>use_partition</code> which are also available in this function with the same defaults.  If you need to fine-tune the trajectory, use <code><a href="https://rdrr.io/pkg/monocle3/man/learn_graph.html" class="external-link">monocle3::learn_graph</a></code> on the cds object first and then run this function to calculate pseudotime.  The graph learning will not be repeated on an object unless <code>force_graph</code> is set to <code>TRUE</code>.</p>
<p>If you just want to look at the trajectory graph and not calculate pseudotime, change <code>calculate_pseudotime</code> to FALSE, or run <code><a href="https://rdrr.io/pkg/monocle3/man/learn_graph.html" class="external-link">monocle3::learn_graph</a></code>.</p>
<p>After the pseudotime values are calculated, they are handled differently than in monocle3.  In this function, they are copied from the hidden CDS slot and made an explicit cell metadata column.  Pseudotime needs a starting point or anchor.  There is no interactive option here as in monocle3.  To identify this starting point, you identify a cell metadata variable and provide it to <code>cluster_variable</code>.  This should identify a cohesive group of cells in UMAP space such as a leiden cluster, louvain cluster or partition.  Then provide a value corresponding to the cluster of interest to <code>cluster_value</code>.  The function will start pseudotime at the cell closest to the graph node in that cluster.  The pseudotime value column will be named automatically as a composite of the <code>cluster_variable</code> and <code>cluster_value</code> parameters.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bb_pseudotime</span><span class="op">(</span></span>
<span>  <span class="va">cds</span>,</span>
<span>  calculate_pseudotime <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">cluster_variable</span>,</span>
<span>  <span class="va">cluster_value</span>,</span>
<span>  use_partition <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  close_loop <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  force_graph <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cds">cds<a class="anchor" aria-label="anchor" href="#arg-cds"></a></dt>
<dd><p>The cell data set object to calculate pseudotime upon.  Does not yet accept seurat objects.</p></dd>


<dt id="arg-calculate-pseudotime">calculate_pseudotime<a class="anchor" aria-label="anchor" href="#arg-calculate-pseudotime"></a></dt>
<dd><p>Logical, whether to calculate the pseudotime dimension.  If false, will only run learn_graph, Default: TRUE</p></dd>


<dt id="arg-cluster-variable">cluster_variable<a class="anchor" aria-label="anchor" href="#arg-cluster-variable"></a></dt>
<dd><p>The cell metadata column from which the pseudotime = 0 cell will be selected.</p></dd>


<dt id="arg-cluster-value">cluster_value<a class="anchor" aria-label="anchor" href="#arg-cluster-value"></a></dt>
<dd><p>The value of cluster_variable that identifies a cluster.  The cell closest to the root node closest to the center of this cluster will have pseudotime of 0.</p></dd>


<dt id="arg-use-partition">use_partition<a class="anchor" aria-label="anchor" href="#arg-use-partition"></a></dt>
<dd><p>Logical; If TRUE, learn_graph will construct trajectories within partitions.  If FALSE, it will connect partitions, Default: TRUE</p></dd>


<dt id="arg-close-loop">close_loop<a class="anchor" aria-label="anchor" href="#arg-close-loop"></a></dt>
<dd><p>Logical; Whether learn_graph will close looping trajectories, Default: TRUE</p></dd>


<dt id="arg-force-graph">force_graph<a class="anchor" aria-label="anchor" href="#arg-force-graph"></a></dt>
<dd><p>Logical; If TRUE, the function will recalculate the graph., Default: FALSE</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A cell data set</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://cli.r-lib.org/reference/cli_div.html" class="external-link">cli_div</a></code>, <code><a href="https://cli.r-lib.org/reference/cli_alert.html" class="external-link">cli_alert</a></code>
<code><a href="https://rdrr.io/pkg/monocle3/man/learn_graph.html" class="external-link">learn_graph</a></code>, <code><a href="https://rdrr.io/pkg/monocle3/man/order_cells.html" class="external-link">order_cells</a></code>, <code><a href="https://rdrr.io/pkg/monocle3/man/pseudotime.html" class="external-link">pseudotime</a></code>
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment-class</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brad Blaser.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

